{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d43146eb-dd1e-44dd-af6c-1ce66157518c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7a7699a-bfa1-4b61-890b-5f28448fb536",
   "metadata": {},
   "source": [
    "# Expression data section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "48f18357-f216-4ff9-a512-29b3abc78b3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_species_id_and_column_name(column_name, reference_columns):\n",
    "    parts = column_name.split(\"_\")\n",
    "    if \"_\".join(parts[2:]) in reference_columns:\n",
    "        return \"_\".join(parts[:2]), \"_\".join(parts[2:])\n",
    "    else:\n",
    "        return parts[0], \"_\".join(parts[1:])\n",
    "\n",
    "\n",
    "def read_xlsx_files_to_dataframe(directory, concatenated_df=None):\n",
    "    # Initialize an empty dataframe if concatenated_df is not provided\n",
    "    if concatenated_df is None:\n",
    "        concatenated_df = pd.DataFrame()\n",
    "\n",
    "    # Get the set of species that have already been read\n",
    "    existing_species = (\n",
    "        set(concatenated_df[\"Species\"])\n",
    "        if \"Species\" in concatenated_df.columns\n",
    "        else set()\n",
    "    )\n",
    "\n",
    "    # Reference columns for species ID extraction\n",
    "    reference_columns = {\n",
    "        col.split(\"_\")[0] for col in concatenated_df.columns if \"TPM\" in col\n",
    "    }\n",
    "    reference_columns.update(\n",
    "        {\n",
    "            \"_\".join(col.split(\"_\")[:2])\n",
    "            for col in concatenated_df.columns\n",
    "            if \"TPM\" in col\n",
    "        }\n",
    "    )\n",
    "\n",
    "    # Iterate over all files in the given directory\n",
    "    for filename in os.listdir(directory):\n",
    "        if filename.endswith(\".xlsx\"):\n",
    "            file_path = os.path.join(directory, filename)\n",
    "\n",
    "            # Remove the extension from the filename\n",
    "            base_filename = os.path.splitext(filename)[0]\n",
    "\n",
    "            # Read the Excel file into a pandas dataframe\n",
    "            df = pd.read_excel(file_path)\n",
    "\n",
    "            # Check if the species in the file is already read\n",
    "            if \"Species\" in df.columns:\n",
    "                species_in_file = set(df[\"Species\"])\n",
    "                if species_in_file.intersection(existing_species):\n",
    "                    print(\n",
    "                        f\"Species {species_in_file} is skipped as it has already been read from {base_filename}.\"\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # Extract 'Species' and 'Region' columns\n",
    "            columns_to_extract = [\"Species\", \"Chromosome\", \"Region\"]\n",
    "\n",
    "            # Identify TPM columns and extract them\n",
    "            tpm_columns = [col for col in df.columns if \"TPM\" in col]\n",
    "\n",
    "            # Create a new dataframe with the required columns\n",
    "            extracted_data = df[columns_to_extract + tpm_columns].copy()\n",
    "\n",
    "            # Extract species id from the TPM columns and rename them\n",
    "            species_ids = []\n",
    "            for col in tpm_columns:\n",
    "                species_id, new_col_name = extract_species_id_and_column_name(\n",
    "                    col, reference_columns\n",
    "                )\n",
    "                extracted_data[new_col_name] = extracted_data[col]\n",
    "                extracted_data.drop(columns=[col], inplace=True)\n",
    "                species_ids.append(species_id)\n",
    "\n",
    "            try:\n",
    "                if len(set(species_ids)) == 1:\n",
    "                    extracted_data[\"Species ID\"] = species_ids[0]\n",
    "                else:\n",
    "                    raise ValueError(\"Multiple species IDs found in one file.\")\n",
    "            except ValueError as e:\n",
    "                print(f\"Error processing file {filename}: {e}\")\n",
    "                extracted_data[\"Species ID\"] = species_ids[0]\n",
    "\n",
    "            # Concatenate the extracted data to the main dataframe\n",
    "            if concatenated_df.empty:\n",
    "                concatenated_df = extracted_data\n",
    "            else:\n",
    "                if set(concatenated_df.columns) == set(extracted_data.columns):\n",
    "                    concatenated_df = pd.concat(\n",
    "                        [concatenated_df, extracted_data], ignore_index=True\n",
    "                    )\n",
    "                    print(f\"Data from {filename} is extracted\")\n",
    "\n",
    "                else:\n",
    "                    unmatched_columns = set(\n",
    "                        concatenated_df.columns\n",
    "                    ).symmetric_difference(set(extracted_data.columns))\n",
    "                    print(\n",
    "                        f\"Column names do not match across the files. Unmatched columns: {unmatched_columns}\"\n",
    "                    )\n",
    "                    return concatenated_df\n",
    "\n",
    "    return concatenated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed5acfd3-9a4f-4a6c-a367-a931301c78f0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data from Shigella flexneri 5a str. M90T.xlsx is extracted\n",
      "Data from Escherichia coli UPEC 536.xlsx is extracted\n",
      "Error processing file Streptococcus suis S10 P 17.xlsx: Multiple species IDs found in one file.\n",
      "Data from Streptococcus suis S10 P 17.xlsx is extracted\n",
      "Data from Acinetobacter baumannii AB5075-UW.xlsx is extracted\n",
      "Data from Klebsiella pneumoniae subsp. pneumoniae MGH 78578.xlsx is extracted\n",
      "Column names do not match across the files. Unmatched columns: {'G27_Tm_3 (GE) - TPM', 'G27_Oxs_2 (GE) - TPM', 'As_3 (GE) - TPM', 'Vic_3 (GE) - TPM', 'G27_Bs_2 (GE) - TPM', 'Li_1 (GE) - TPM', 'G27_Nd_1 (GE) - TPM', 'Li_3 (GE) - TPM', 'G27_Oxs_3 (GE) - TPM', 'G27_Tm_1 (GE) - TPM', 'As_2 (GE) - TPM', 'G27_Ns_1 (GE) - TPM', 'G27_Bs_1 (GE) - TPM', 'Oxs_3 (GE) - TPM', 'Sp_2 (GE) - TPM', 'G27_Li_1 (GE) - TPM', 'G27_Bs_3 (GE) - TPM', 'G27_Nd_3 (GE) - TPM', 'G27_As_3 (GE) - TPM', 'G27_Mig_2 (GE) - TPM', 'Oxs_2 (GE) - TPM', 'Mig_3 (GE) - TPM', 'G27_Li_3 (GE) - TPM', 'Vic_2 (GE) - TPM', 'Ns_3 (GE) - TPM', 'Bs_1 (GE) - TPM', 'Oss_1 (GE) - TPM', 'Ns_1 (GE) - TPM', 'G27_Vic_2 (GE) - TPM', 'G27_As_2 (GE) - TPM', 'G27_Ns_3 (GE) - TPM', 'G27_Oss_3 (GE) - TPM', 'G27_Mig_1 (GE) - TPM', 'Nd_2 (GE) - TPM', 'G27_Nd_2 (GE) - TPM', 'Tm_3 (GE) - TPM', 'Bs_3 (GE) - TPM', 'Sp_1 (GE) - TPM', 'G27_Tm_2 (GE) - TPM', 'Vic_1 (GE) - TPM', 'Tm_1 (GE) - TPM', 'Ctrl_2 (GE) - TPM', 'G27_Li_2 (GE) - TPM', 'G27_Ns_2 (GE) - TPM', 'Bs_2 (GE) - TPM', 'G27_As_1 (GE) - TPM', 'G27_Vic_3 (GE) - TPM', 'G27_Oss_2 (GE) - TPM', 'Ctrl_3 (GE) - TPM', 'As_1 (GE) - TPM', 'G27_Ctrl_1 (GE) - TPM', 'G27_Ctrl_3 (GE) - TPM', 'G27_Oxs_1 (GE) - TPM', 'Oxs_1 (GE) - TPM', 'Tm_2 (GE) - TPM', 'G27_Ctrl_2 (GE) - TPM', 'Nd_1 (GE) - TPM', 'Li_2 (GE) - TPM', 'G27_Sp_2 (GE) - TPM', 'Mig_1 (GE) - TPM', 'Mig_2 (GE) - TPM', 'Sp_3 (GE) - TPM', 'Oss_3 (GE) - TPM', 'G27_Mig_3 (GE) - TPM', 'G27_Vic_1 (GE) - TPM', 'Ctrl_1 (GE) - TPM', 'G27_Sp_1 (GE) - TPM', 'Ns_2 (GE) - TPM', 'Oss_2 (GE) - TPM', 'Nd_3 (GE) - TPM', 'G27_Oss_1 (GE) - TPM', 'G27_Sp_3 (GE) - TPM'}\n",
      "                                                 Species Chromosome  \\\n",
      "0           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
      "1           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
      "2           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
      "3           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
      "4           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
      "...                                                  ...        ...   \n",
      "24127  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
      "24128  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
      "24129  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
      "24130  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
      "24131  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
      "\n",
      "                       Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
      "0        complement(235..402)         0.000000         0.000000   \n",
      "1        complement(372..806)       393.166189       311.937046   \n",
      "2       complement(816..2210)       162.400143       171.281001   \n",
      "3      complement(2271..3896)       720.122251       825.256114   \n",
      "4      complement(3899..4156)       132.002479       371.588303   \n",
      "...                       ...              ...              ...   \n",
      "24127              1433..1663      2196.775545      2257.918323   \n",
      "24128              1660..1998      2181.385479      1981.081622   \n",
      "24129  complement(2287..2544)       107.262956       165.195837   \n",
      "24130              2667..2858        59.660108        97.389121   \n",
      "24131              2858..3259       464.105365       520.751232   \n",
      "\n",
      "       As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
      "0             0.000000         0.000000         0.000000         0.000000   \n",
      "1           383.829258       377.730614       273.496217       331.362458   \n",
      "2           174.959500       261.605512       228.250776       233.104257   \n",
      "3           860.384381       560.678176       683.429275       731.909515   \n",
      "4           244.840701       150.111163       232.967463       208.688639   \n",
      "...                ...              ...              ...              ...   \n",
      "24127      2155.878280      1635.286545      2357.383900      1929.630652   \n",
      "24128      1666.816029      1746.367340      1849.783568      1862.333827   \n",
      "24129       160.945638       268.944256       336.767008       251.302784   \n",
      "24130        74.907776        46.934265        44.546468        41.170402   \n",
      "24131       461.618791       264.223870       379.004922       324.158181   \n",
      "\n",
      "       Ctrl_1 (GE) - TPM  ...  Sp_1 (GE) - TPM  Sp_2 (GE) - TPM  \\\n",
      "0               0.000000  ...         0.000000         0.000000   \n",
      "1             315.921082  ...       302.104930       175.084374   \n",
      "2             154.689875  ...       203.637405       145.679966   \n",
      "3             554.730307  ...       276.626761       442.196629   \n",
      "4             205.699493  ...       170.299063       242.590426   \n",
      "...                  ...  ...              ...              ...   \n",
      "24127        2260.813993  ...     12814.854657      7358.802631   \n",
      "24128        1403.000191  ...     11653.963228      9174.082484   \n",
      "24129         306.821038  ...        68.457070        47.994576   \n",
      "24130          47.496680  ...       188.819912       283.767930   \n",
      "24131         269.647255  ...       647.465139       554.444802   \n",
      "\n",
      "       Sp_3 (GE) - TPM  Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  \\\n",
      "0             0.000000         8.392188         0.000000         0.000000   \n",
      "1           351.995974       385.693391       383.276940       406.241639   \n",
      "2           220.608060       260.216487       177.158771       215.282147   \n",
      "3           473.234066       327.976622       676.676655       632.319698   \n",
      "4           153.183540       128.419995       108.867940       232.579308   \n",
      "...                ...              ...              ...              ...   \n",
      "24127     10632.870032     12801.790715     10104.657626     10909.565036   \n",
      "24128      8283.044866      4301.461428      4804.391257      3884.423218   \n",
      "24129        67.531582       250.489120       247.530226       322.815816   \n",
      "24130        90.745563       208.279627       163.870817       182.534512   \n",
      "24131       511.894292       865.093206       878.869355       807.683394   \n",
      "\n",
      "       Vic_1 (GE) - TPM  Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \n",
      "0              0.000000          0.000000          0.000000      Vibrio  \n",
      "1            276.773111        178.371963        223.310216      Vibrio  \n",
      "2            162.996568         97.415072        128.631521      Vibrio  \n",
      "3            622.953497        648.744407        603.585147      Vibrio  \n",
      "4            223.051153        367.948661        124.616224      Vibrio  \n",
      "...                 ...               ...               ...         ...  \n",
      "24127       9206.009642       6822.881422       4465.778497       KLEBS  \n",
      "24128       3057.111744       3325.739216       3063.158487       KLEBS  \n",
      "24129        209.147013        255.838672        283.460097       KLEBS  \n",
      "24130        105.147091        141.333099        141.038356       KLEBS  \n",
      "24131        691.603197        745.262717        564.096144       KLEBS  \n",
      "\n",
      "[24132 rows x 40 columns]\n"
     ]
    }
   ],
   "source": [
    "# Reads expression data and creates a dataframe.\n",
    "directory = f\"{os.getcwd()}/data/data_expression\"\n",
    "final_df = read_xlsx_files_to_dataframe(directory)\n",
    "print(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ce933e0-aa0e-4ff2-b809-d47fe5f0a4ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>As_1 (GE) - TPM</th>\n",
       "      <th>As_2 (GE) - TPM</th>\n",
       "      <th>As_3 (GE) - TPM</th>\n",
       "      <th>Bs_1 (GE) - TPM</th>\n",
       "      <th>Bs_2 (GE) - TPM</th>\n",
       "      <th>Bs_3 (GE) - TPM</th>\n",
       "      <th>Ctrl_1 (GE) - TPM</th>\n",
       "      <th>...</th>\n",
       "      <th>Sp_1 (GE) - TPM</th>\n",
       "      <th>Sp_2 (GE) - TPM</th>\n",
       "      <th>Sp_3 (GE) - TPM</th>\n",
       "      <th>Tm_1 (GE) - TPM</th>\n",
       "      <th>Tm_2 (GE) - TPM</th>\n",
       "      <th>Tm_3 (GE) - TPM</th>\n",
       "      <th>Vic_1 (GE) - TPM</th>\n",
       "      <th>Vic_2 (GE) - TPM</th>\n",
       "      <th>Vic_3 (GE) - TPM</th>\n",
       "      <th>Species ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(235..402)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.392188</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Vibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(372..806)</td>\n",
       "      <td>393.166189</td>\n",
       "      <td>311.937046</td>\n",
       "      <td>383.829258</td>\n",
       "      <td>377.730614</td>\n",
       "      <td>273.496217</td>\n",
       "      <td>331.362458</td>\n",
       "      <td>315.921082</td>\n",
       "      <td>...</td>\n",
       "      <td>302.104930</td>\n",
       "      <td>175.084374</td>\n",
       "      <td>351.995974</td>\n",
       "      <td>385.693391</td>\n",
       "      <td>383.276940</td>\n",
       "      <td>406.241639</td>\n",
       "      <td>276.773111</td>\n",
       "      <td>178.371963</td>\n",
       "      <td>223.310216</td>\n",
       "      <td>Vibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(816..2210)</td>\n",
       "      <td>162.400143</td>\n",
       "      <td>171.281001</td>\n",
       "      <td>174.959500</td>\n",
       "      <td>261.605512</td>\n",
       "      <td>228.250776</td>\n",
       "      <td>233.104257</td>\n",
       "      <td>154.689875</td>\n",
       "      <td>...</td>\n",
       "      <td>203.637405</td>\n",
       "      <td>145.679966</td>\n",
       "      <td>220.608060</td>\n",
       "      <td>260.216487</td>\n",
       "      <td>177.158771</td>\n",
       "      <td>215.282147</td>\n",
       "      <td>162.996568</td>\n",
       "      <td>97.415072</td>\n",
       "      <td>128.631521</td>\n",
       "      <td>Vibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(2271..3896)</td>\n",
       "      <td>720.122251</td>\n",
       "      <td>825.256114</td>\n",
       "      <td>860.384381</td>\n",
       "      <td>560.678176</td>\n",
       "      <td>683.429275</td>\n",
       "      <td>731.909515</td>\n",
       "      <td>554.730307</td>\n",
       "      <td>...</td>\n",
       "      <td>276.626761</td>\n",
       "      <td>442.196629</td>\n",
       "      <td>473.234066</td>\n",
       "      <td>327.976622</td>\n",
       "      <td>676.676655</td>\n",
       "      <td>632.319698</td>\n",
       "      <td>622.953497</td>\n",
       "      <td>648.744407</td>\n",
       "      <td>603.585147</td>\n",
       "      <td>Vibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(3899..4156)</td>\n",
       "      <td>132.002479</td>\n",
       "      <td>371.588303</td>\n",
       "      <td>244.840701</td>\n",
       "      <td>150.111163</td>\n",
       "      <td>232.967463</td>\n",
       "      <td>208.688639</td>\n",
       "      <td>205.699493</td>\n",
       "      <td>...</td>\n",
       "      <td>170.299063</td>\n",
       "      <td>242.590426</td>\n",
       "      <td>153.183540</td>\n",
       "      <td>128.419995</td>\n",
       "      <td>108.867940</td>\n",
       "      <td>232.579308</td>\n",
       "      <td>223.051153</td>\n",
       "      <td>367.948661</td>\n",
       "      <td>124.616224</td>\n",
       "      <td>Vibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24127</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>1433..1663</td>\n",
       "      <td>2196.775545</td>\n",
       "      <td>2257.918323</td>\n",
       "      <td>2155.878280</td>\n",
       "      <td>1635.286545</td>\n",
       "      <td>2357.383900</td>\n",
       "      <td>1929.630652</td>\n",
       "      <td>2260.813993</td>\n",
       "      <td>...</td>\n",
       "      <td>12814.854657</td>\n",
       "      <td>7358.802631</td>\n",
       "      <td>10632.870032</td>\n",
       "      <td>12801.790715</td>\n",
       "      <td>10104.657626</td>\n",
       "      <td>10909.565036</td>\n",
       "      <td>9206.009642</td>\n",
       "      <td>6822.881422</td>\n",
       "      <td>4465.778497</td>\n",
       "      <td>KLEBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24128</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>1660..1998</td>\n",
       "      <td>2181.385479</td>\n",
       "      <td>1981.081622</td>\n",
       "      <td>1666.816029</td>\n",
       "      <td>1746.367340</td>\n",
       "      <td>1849.783568</td>\n",
       "      <td>1862.333827</td>\n",
       "      <td>1403.000191</td>\n",
       "      <td>...</td>\n",
       "      <td>11653.963228</td>\n",
       "      <td>9174.082484</td>\n",
       "      <td>8283.044866</td>\n",
       "      <td>4301.461428</td>\n",
       "      <td>4804.391257</td>\n",
       "      <td>3884.423218</td>\n",
       "      <td>3057.111744</td>\n",
       "      <td>3325.739216</td>\n",
       "      <td>3063.158487</td>\n",
       "      <td>KLEBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24129</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>complement(2287..2544)</td>\n",
       "      <td>107.262956</td>\n",
       "      <td>165.195837</td>\n",
       "      <td>160.945638</td>\n",
       "      <td>268.944256</td>\n",
       "      <td>336.767008</td>\n",
       "      <td>251.302784</td>\n",
       "      <td>306.821038</td>\n",
       "      <td>...</td>\n",
       "      <td>68.457070</td>\n",
       "      <td>47.994576</td>\n",
       "      <td>67.531582</td>\n",
       "      <td>250.489120</td>\n",
       "      <td>247.530226</td>\n",
       "      <td>322.815816</td>\n",
       "      <td>209.147013</td>\n",
       "      <td>255.838672</td>\n",
       "      <td>283.460097</td>\n",
       "      <td>KLEBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24130</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>2667..2858</td>\n",
       "      <td>59.660108</td>\n",
       "      <td>97.389121</td>\n",
       "      <td>74.907776</td>\n",
       "      <td>46.934265</td>\n",
       "      <td>44.546468</td>\n",
       "      <td>41.170402</td>\n",
       "      <td>47.496680</td>\n",
       "      <td>...</td>\n",
       "      <td>188.819912</td>\n",
       "      <td>283.767930</td>\n",
       "      <td>90.745563</td>\n",
       "      <td>208.279627</td>\n",
       "      <td>163.870817</td>\n",
       "      <td>182.534512</td>\n",
       "      <td>105.147091</td>\n",
       "      <td>141.333099</td>\n",
       "      <td>141.038356</td>\n",
       "      <td>KLEBS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24131</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>2858..3259</td>\n",
       "      <td>464.105365</td>\n",
       "      <td>520.751232</td>\n",
       "      <td>461.618791</td>\n",
       "      <td>264.223870</td>\n",
       "      <td>379.004922</td>\n",
       "      <td>324.158181</td>\n",
       "      <td>269.647255</td>\n",
       "      <td>...</td>\n",
       "      <td>647.465139</td>\n",
       "      <td>554.444802</td>\n",
       "      <td>511.894292</td>\n",
       "      <td>865.093206</td>\n",
       "      <td>878.869355</td>\n",
       "      <td>807.683394</td>\n",
       "      <td>691.603197</td>\n",
       "      <td>745.262717</td>\n",
       "      <td>564.096144</td>\n",
       "      <td>KLEBS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24132 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Species Chromosome  \\\n",
       "0           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "1           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "2           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "3           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "4           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "...                                                  ...        ...   \n",
       "24127  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "24128  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "24129  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "24130  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "24131  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "\n",
       "                       Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
       "0        complement(235..402)         0.000000         0.000000   \n",
       "1        complement(372..806)       393.166189       311.937046   \n",
       "2       complement(816..2210)       162.400143       171.281001   \n",
       "3      complement(2271..3896)       720.122251       825.256114   \n",
       "4      complement(3899..4156)       132.002479       371.588303   \n",
       "...                       ...              ...              ...   \n",
       "24127              1433..1663      2196.775545      2257.918323   \n",
       "24128              1660..1998      2181.385479      1981.081622   \n",
       "24129  complement(2287..2544)       107.262956       165.195837   \n",
       "24130              2667..2858        59.660108        97.389121   \n",
       "24131              2858..3259       464.105365       520.751232   \n",
       "\n",
       "       As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
       "0             0.000000         0.000000         0.000000         0.000000   \n",
       "1           383.829258       377.730614       273.496217       331.362458   \n",
       "2           174.959500       261.605512       228.250776       233.104257   \n",
       "3           860.384381       560.678176       683.429275       731.909515   \n",
       "4           244.840701       150.111163       232.967463       208.688639   \n",
       "...                ...              ...              ...              ...   \n",
       "24127      2155.878280      1635.286545      2357.383900      1929.630652   \n",
       "24128      1666.816029      1746.367340      1849.783568      1862.333827   \n",
       "24129       160.945638       268.944256       336.767008       251.302784   \n",
       "24130        74.907776        46.934265        44.546468        41.170402   \n",
       "24131       461.618791       264.223870       379.004922       324.158181   \n",
       "\n",
       "       Ctrl_1 (GE) - TPM  ...  Sp_1 (GE) - TPM  Sp_2 (GE) - TPM  \\\n",
       "0               0.000000  ...         0.000000         0.000000   \n",
       "1             315.921082  ...       302.104930       175.084374   \n",
       "2             154.689875  ...       203.637405       145.679966   \n",
       "3             554.730307  ...       276.626761       442.196629   \n",
       "4             205.699493  ...       170.299063       242.590426   \n",
       "...                  ...  ...              ...              ...   \n",
       "24127        2260.813993  ...     12814.854657      7358.802631   \n",
       "24128        1403.000191  ...     11653.963228      9174.082484   \n",
       "24129         306.821038  ...        68.457070        47.994576   \n",
       "24130          47.496680  ...       188.819912       283.767930   \n",
       "24131         269.647255  ...       647.465139       554.444802   \n",
       "\n",
       "       Sp_3 (GE) - TPM  Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  \\\n",
       "0             0.000000         8.392188         0.000000         0.000000   \n",
       "1           351.995974       385.693391       383.276940       406.241639   \n",
       "2           220.608060       260.216487       177.158771       215.282147   \n",
       "3           473.234066       327.976622       676.676655       632.319698   \n",
       "4           153.183540       128.419995       108.867940       232.579308   \n",
       "...                ...              ...              ...              ...   \n",
       "24127     10632.870032     12801.790715     10104.657626     10909.565036   \n",
       "24128      8283.044866      4301.461428      4804.391257      3884.423218   \n",
       "24129        67.531582       250.489120       247.530226       322.815816   \n",
       "24130        90.745563       208.279627       163.870817       182.534512   \n",
       "24131       511.894292       865.093206       878.869355       807.683394   \n",
       "\n",
       "       Vic_1 (GE) - TPM  Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \n",
       "0              0.000000          0.000000          0.000000      Vibrio  \n",
       "1            276.773111        178.371963        223.310216      Vibrio  \n",
       "2            162.996568         97.415072        128.631521      Vibrio  \n",
       "3            622.953497        648.744407        603.585147      Vibrio  \n",
       "4            223.051153        367.948661        124.616224      Vibrio  \n",
       "...                 ...               ...               ...         ...  \n",
       "24127       9206.009642       6822.881422       4465.778497       KLEBS  \n",
       "24128       3057.111744       3325.739216       3063.158487       KLEBS  \n",
       "24129        209.147013        255.838672        283.460097       KLEBS  \n",
       "24130        105.147091        141.333099        141.038356       KLEBS  \n",
       "24131        691.603197        745.262717        564.096144       KLEBS  \n",
       "\n",
       "[24132 rows x 40 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "98c6d709-bc23-4ab9-85cd-e13fe1588fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv(f\"{os.getcwd()}/data/expression_final_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f0294b83-f340-4c4c-8d95-0df59f680daf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species:  6\n",
      "{'Vibrio cholerae O1 biovar El Tor str. N16961', 'Shigella flexneri 5a str. M90T', 'Escherichia coli UPEC 536', 'Klebsiella pneumoniae subsp. pneumoniae MGH 78578', 'Acinetobacter baumannii AB5075-UW', 'Streptococcus suis S10 -P1/7'}\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of species: \", len(set(final_df[\"Species\"].tolist())))\n",
    "print(set(final_df[\"Species\"].tolist()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3e2724a-adde-4dc1-9704-164d48d9e2ed",
   "metadata": {},
   "source": [
    "# upstream_sequences section (Move after you get all the species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e3867fe2-b4fa-46bc-bb20-059c5efceb84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions to process sequence data\n",
    "def preprocess_tsv(tsv_path):\n",
    "    # Read the TSV file into a pandas dataframe\n",
    "    tsv_df = pd.read_csv(tsv_path, sep=\"\\t\")\n",
    "\n",
    "    # Drop the .csv extension from the 'csv' column and rename it to 'Species'\n",
    "    tsv_df[\"Species\"] = tsv_df[\"csv\"].str.replace(\".csv\", \"\", regex=False)\n",
    "\n",
    "    # Rename the 'contig' column to 'Chromosome'\n",
    "    tsv_df.rename(columns={\"contig\": \"Chromosome\", \"region\": \"Region\"}, inplace=True)\n",
    "\n",
    "    # Drop the original 'csv' column\n",
    "    tsv_df.drop(columns=[\"csv\"], inplace=True)\n",
    "\n",
    "    # Replace specific species names\n",
    "    tsv_df[\"Species\"] = tsv_df[\"Species\"].replace(\n",
    "        {\n",
    "            \"Salmonella enterica subsp. enterica serovar TyphimuriumSL1344\": \"Salmonella enterica subsp. enterica serovar Typhimurium SL1344\",\n",
    "            \"Escherichia coli EPEC 0127 H6 E2348 69\": \"Escherichia coli EPEC 0127:H6 E2348/69\",\n",
    "            \"Streptococcus suis S10 P 17\": \"Streptococcus suis S10 -P1/7\",\n",
    "        }\n",
    "    )\n",
    "\n",
    "    return tsv_df\n",
    "\n",
    "\n",
    "def merge_dataframes(main_df, tsv_df):\n",
    "    # Check if the two datasets have the same species\n",
    "    species_diff_1 = set(tsv_df[\"Species\"]).difference(set(main_df[\"Species\"]))\n",
    "    species_diff_2 = set(main_df[\"Species\"]).difference(set(tsv_df[\"Species\"]))\n",
    "\n",
    "    if species_diff_1 or species_diff_2:\n",
    "        print(f\"Species present in TSV but not in main dataframe: {species_diff_1}\")\n",
    "        print(f\"Species present in main dataframe but not in TSV: {species_diff_2}\")\n",
    "\n",
    "    # Perform an inner merge on 'Species', 'Chromosome', and 'Region' columns\n",
    "    # What other merge can we do here!!\n",
    "    merged_df = pd.merge(\n",
    "        main_df, tsv_df, on=[\"Species\", \"Chromosome\", \"Region\"], how=\"inner\"\n",
    "    )\n",
    "\n",
    "    return merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "89b402ce-6af9-4388-961e-343abfc1161e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read and preprocess the TSV file\n",
    "tsv_path = f\"{os.getcwd()}/data/data_sequences_upstream/upstream_sequences.tsv\"\n",
    "sequence_df = preprocess_tsv(tsv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7b20ac7a-48d4-49f5-87f1-0e763c90c895",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>upstream200</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(235..402)</td>\n",
       "      <td>CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(372..806)</td>\n",
       "      <td>ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(816..2210)</td>\n",
       "      <td>TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(2271..3896)</td>\n",
       "      <td>TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(3899..4156)</td>\n",
       "      <td>CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96406</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3393934..3395274)</td>\n",
       "      <td>CAGCACCAGCTGATCCAATGCAAGCTAAGGTAATGATGTTTTTACC...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96407</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3395275..3396945)</td>\n",
       "      <td>ATCAGTATTTTATTAGCCCTTTGATAACACCATGTTGTCGCTATTA...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96408</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3396955..3397200)</td>\n",
       "      <td>AATAAGTTAGGCTATGCACGCCTTGGTTTAGCATTGTCAAAAAAAA...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96409</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3397167..3397355)</td>\n",
       "      <td>TAAAAAGACGTCGTGCTAAAGGTCGTAAGCGTTTATCTGCCTAAGT...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96410</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3397512..3397646)</td>\n",
       "      <td>CGACCTGTAAATAAAAAATTACACAAAAAATAATCAGTTATTCAAA...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96411 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome                        Region  \\\n",
       "0      NC_002505          complement(235..402)   \n",
       "1      NC_002505          complement(372..806)   \n",
       "2      NC_002505         complement(816..2210)   \n",
       "3      NC_002505        complement(2271..3896)   \n",
       "4      NC_002505        complement(3899..4156)   \n",
       "...          ...                           ...   \n",
       "96406  NC_002942  complement(3393934..3395274)   \n",
       "96407  NC_002942  complement(3395275..3396945)   \n",
       "96408  NC_002942  complement(3396955..3397200)   \n",
       "96409  NC_002942  complement(3397167..3397355)   \n",
       "96410  NC_002942  complement(3397512..3397646)   \n",
       "\n",
       "                                             upstream200  \\\n",
       "0      CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...   \n",
       "1      ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...   \n",
       "2      TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...   \n",
       "3      TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...   \n",
       "4      CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...   \n",
       "...                                                  ...   \n",
       "96406  CAGCACCAGCTGATCCAATGCAAGCTAAGGTAATGATGTTTTTACC...   \n",
       "96407  ATCAGTATTTTATTAGCCCTTTGATAACACCATGTTGTCGCTATTA...   \n",
       "96408  AATAAGTTAGGCTATGCACGCCTTGGTTTAGCATTGTCAAAAAAAA...   \n",
       "96409  TAAAAAGACGTCGTGCTAAAGGTCGTAAGCGTTTATCTGCCTAAGT...   \n",
       "96410  CGACCTGTAAATAAAAAATTACACAAAAAATAATCAGTTATTCAAA...   \n",
       "\n",
       "                                                 Species  \n",
       "0           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "1           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "2           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "3           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "4           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "...                                                  ...  \n",
       "96406  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96407  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96408  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96409  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96410  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "\n",
       "[96411 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4d35cf33-bc52-4808-9c18-66023e79a992",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species present in TSV but not in main dataframe: {'Enterococcus faecalis OG1RF', 'Legionella pneumophila subsp. pneumophila Philadelphia 1', 'Escherichia coli ETEC H10407', 'Streptococcus pneumoniae D39', 'Helicobacter pylori G27', 'Borrelia burgdorferi B31', 'Staphylococcus\\xa0aureus MRSA252', 'Burkholderia pseudomallei K96243', 'Streptococcus agalactiae NEM316', 'Haemophilus influenzae 86-028NP', 'Campylobacter jejuni subsp. jejuni 81-176', 'Listeria monocytogenes EGD-e', 'Neisseria meningitidis serogroup C FAM18', 'Francisella tularensis subsp. holarctica FSC200', 'Mycobacterium tuberculosis H37Ra', 'Streptococcus pyogenes 5448', 'Aggregatibacter actinomycetemcomitans D7S-1', 'Escherichia coli EPEC 0127:H6 E2348/69', 'Staphylococcus\\xa0epidermidis 1457', 'Salmonella enterica subsp. enterica serovar Typhimurium SL1344', 'Achromobacter xylosoxidans SOLR10', 'Staphylococcus\\xa0aureus MSSA476', 'Helicobacter pylori J99', 'Neisseria gonorrhoeae FA 1090', 'Pseudomonas aeruginosa PAO1'}\n",
      "Species present in main dataframe but not in TSV: {'Shigella flexneri 5a str. M90T'}\n"
     ]
    }
   ],
   "source": [
    "merged_df = merge_dataframes(final_df, sequence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8dc7950f-e119-44ad-9abd-f7ba476e0eb7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>As_1 (GE) - TPM</th>\n",
       "      <th>As_2 (GE) - TPM</th>\n",
       "      <th>As_3 (GE) - TPM</th>\n",
       "      <th>Bs_1 (GE) - TPM</th>\n",
       "      <th>Bs_2 (GE) - TPM</th>\n",
       "      <th>Bs_3 (GE) - TPM</th>\n",
       "      <th>Ctrl_1 (GE) - TPM</th>\n",
       "      <th>...</th>\n",
       "      <th>Sp_2 (GE) - TPM</th>\n",
       "      <th>Sp_3 (GE) - TPM</th>\n",
       "      <th>Tm_1 (GE) - TPM</th>\n",
       "      <th>Tm_2 (GE) - TPM</th>\n",
       "      <th>Tm_3 (GE) - TPM</th>\n",
       "      <th>Vic_1 (GE) - TPM</th>\n",
       "      <th>Vic_2 (GE) - TPM</th>\n",
       "      <th>Vic_3 (GE) - TPM</th>\n",
       "      <th>Species ID</th>\n",
       "      <th>upstream200</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(235..402)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.392188</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(372..806)</td>\n",
       "      <td>393.166189</td>\n",
       "      <td>311.937046</td>\n",
       "      <td>383.829258</td>\n",
       "      <td>377.730614</td>\n",
       "      <td>273.496217</td>\n",
       "      <td>331.362458</td>\n",
       "      <td>315.921082</td>\n",
       "      <td>...</td>\n",
       "      <td>175.084374</td>\n",
       "      <td>351.995974</td>\n",
       "      <td>385.693391</td>\n",
       "      <td>383.276940</td>\n",
       "      <td>406.241639</td>\n",
       "      <td>276.773111</td>\n",
       "      <td>178.371963</td>\n",
       "      <td>223.310216</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(816..2210)</td>\n",
       "      <td>162.400143</td>\n",
       "      <td>171.281001</td>\n",
       "      <td>174.959500</td>\n",
       "      <td>261.605512</td>\n",
       "      <td>228.250776</td>\n",
       "      <td>233.104257</td>\n",
       "      <td>154.689875</td>\n",
       "      <td>...</td>\n",
       "      <td>145.679966</td>\n",
       "      <td>220.608060</td>\n",
       "      <td>260.216487</td>\n",
       "      <td>177.158771</td>\n",
       "      <td>215.282147</td>\n",
       "      <td>162.996568</td>\n",
       "      <td>97.415072</td>\n",
       "      <td>128.631521</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(2271..3896)</td>\n",
       "      <td>720.122251</td>\n",
       "      <td>825.256114</td>\n",
       "      <td>860.384381</td>\n",
       "      <td>560.678176</td>\n",
       "      <td>683.429275</td>\n",
       "      <td>731.909515</td>\n",
       "      <td>554.730307</td>\n",
       "      <td>...</td>\n",
       "      <td>442.196629</td>\n",
       "      <td>473.234066</td>\n",
       "      <td>327.976622</td>\n",
       "      <td>676.676655</td>\n",
       "      <td>632.319698</td>\n",
       "      <td>622.953497</td>\n",
       "      <td>648.744407</td>\n",
       "      <td>603.585147</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(3899..4156)</td>\n",
       "      <td>132.002479</td>\n",
       "      <td>371.588303</td>\n",
       "      <td>244.840701</td>\n",
       "      <td>150.111163</td>\n",
       "      <td>232.967463</td>\n",
       "      <td>208.688639</td>\n",
       "      <td>205.699493</td>\n",
       "      <td>...</td>\n",
       "      <td>242.590426</td>\n",
       "      <td>153.183540</td>\n",
       "      <td>128.419995</td>\n",
       "      <td>108.867940</td>\n",
       "      <td>232.579308</td>\n",
       "      <td>223.051153</td>\n",
       "      <td>367.948661</td>\n",
       "      <td>124.616224</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19821</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>1433..1663</td>\n",
       "      <td>2196.775545</td>\n",
       "      <td>2257.918323</td>\n",
       "      <td>2155.878280</td>\n",
       "      <td>1635.286545</td>\n",
       "      <td>2357.383900</td>\n",
       "      <td>1929.630652</td>\n",
       "      <td>2260.813993</td>\n",
       "      <td>...</td>\n",
       "      <td>7358.802631</td>\n",
       "      <td>10632.870032</td>\n",
       "      <td>12801.790715</td>\n",
       "      <td>10104.657626</td>\n",
       "      <td>10909.565036</td>\n",
       "      <td>9206.009642</td>\n",
       "      <td>6822.881422</td>\n",
       "      <td>4465.778497</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>GTAAATCAATTTGTTATCGCCACTTAGTTAAAAAAAATTTTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19822</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>1660..1998</td>\n",
       "      <td>2181.385479</td>\n",
       "      <td>1981.081622</td>\n",
       "      <td>1666.816029</td>\n",
       "      <td>1746.367340</td>\n",
       "      <td>1849.783568</td>\n",
       "      <td>1862.333827</td>\n",
       "      <td>1403.000191</td>\n",
       "      <td>...</td>\n",
       "      <td>9174.082484</td>\n",
       "      <td>8283.044866</td>\n",
       "      <td>4301.461428</td>\n",
       "      <td>4804.391257</td>\n",
       "      <td>3884.423218</td>\n",
       "      <td>3057.111744</td>\n",
       "      <td>3325.739216</td>\n",
       "      <td>3063.158487</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>AATGGCGCTGCTTTGCGTATCCCGCAGCCGTTCATGAAGCAGCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19823</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>complement(2287..2544)</td>\n",
       "      <td>107.262956</td>\n",
       "      <td>165.195837</td>\n",
       "      <td>160.945638</td>\n",
       "      <td>268.944256</td>\n",
       "      <td>336.767008</td>\n",
       "      <td>251.302784</td>\n",
       "      <td>306.821038</td>\n",
       "      <td>...</td>\n",
       "      <td>47.994576</td>\n",
       "      <td>67.531582</td>\n",
       "      <td>250.489120</td>\n",
       "      <td>247.530226</td>\n",
       "      <td>322.815816</td>\n",
       "      <td>209.147013</td>\n",
       "      <td>255.838672</td>\n",
       "      <td>283.460097</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>AAGCTTCTCCAGCAGGAGCAGAGACTGATGCTGGATGAATTTAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19824</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>2667..2858</td>\n",
       "      <td>59.660108</td>\n",
       "      <td>97.389121</td>\n",
       "      <td>74.907776</td>\n",
       "      <td>46.934265</td>\n",
       "      <td>44.546468</td>\n",
       "      <td>41.170402</td>\n",
       "      <td>47.496680</td>\n",
       "      <td>...</td>\n",
       "      <td>283.767930</td>\n",
       "      <td>90.745563</td>\n",
       "      <td>208.279627</td>\n",
       "      <td>163.870817</td>\n",
       "      <td>182.534512</td>\n",
       "      <td>105.147091</td>\n",
       "      <td>141.333099</td>\n",
       "      <td>141.038356</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>GATGCTTGTCAGGGGGGCGGAGCCTATGGAAAAACGGCCGTTGTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19825</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>2858..3259</td>\n",
       "      <td>464.105365</td>\n",
       "      <td>520.751232</td>\n",
       "      <td>461.618791</td>\n",
       "      <td>264.223870</td>\n",
       "      <td>379.004922</td>\n",
       "      <td>324.158181</td>\n",
       "      <td>269.647255</td>\n",
       "      <td>...</td>\n",
       "      <td>554.444802</td>\n",
       "      <td>511.894292</td>\n",
       "      <td>865.093206</td>\n",
       "      <td>878.869355</td>\n",
       "      <td>807.683394</td>\n",
       "      <td>691.603197</td>\n",
       "      <td>745.262717</td>\n",
       "      <td>564.096144</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>AGGCTTCTTATGACGAAACAACAGAAAACAGCGCTGAACATGGCTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19826 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Species Chromosome  \\\n",
       "0           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "1           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "2           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "3           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "4           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "...                                                  ...        ...   \n",
       "19821  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19822  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19823  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19824  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19825  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "\n",
       "                       Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
       "0        complement(235..402)         0.000000         0.000000   \n",
       "1        complement(372..806)       393.166189       311.937046   \n",
       "2       complement(816..2210)       162.400143       171.281001   \n",
       "3      complement(2271..3896)       720.122251       825.256114   \n",
       "4      complement(3899..4156)       132.002479       371.588303   \n",
       "...                       ...              ...              ...   \n",
       "19821              1433..1663      2196.775545      2257.918323   \n",
       "19822              1660..1998      2181.385479      1981.081622   \n",
       "19823  complement(2287..2544)       107.262956       165.195837   \n",
       "19824              2667..2858        59.660108        97.389121   \n",
       "19825              2858..3259       464.105365       520.751232   \n",
       "\n",
       "       As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
       "0             0.000000         0.000000         0.000000         0.000000   \n",
       "1           383.829258       377.730614       273.496217       331.362458   \n",
       "2           174.959500       261.605512       228.250776       233.104257   \n",
       "3           860.384381       560.678176       683.429275       731.909515   \n",
       "4           244.840701       150.111163       232.967463       208.688639   \n",
       "...                ...              ...              ...              ...   \n",
       "19821      2155.878280      1635.286545      2357.383900      1929.630652   \n",
       "19822      1666.816029      1746.367340      1849.783568      1862.333827   \n",
       "19823       160.945638       268.944256       336.767008       251.302784   \n",
       "19824        74.907776        46.934265        44.546468        41.170402   \n",
       "19825       461.618791       264.223870       379.004922       324.158181   \n",
       "\n",
       "       Ctrl_1 (GE) - TPM  ...  Sp_2 (GE) - TPM  Sp_3 (GE) - TPM  \\\n",
       "0               0.000000  ...         0.000000         0.000000   \n",
       "1             315.921082  ...       175.084374       351.995974   \n",
       "2             154.689875  ...       145.679966       220.608060   \n",
       "3             554.730307  ...       442.196629       473.234066   \n",
       "4             205.699493  ...       242.590426       153.183540   \n",
       "...                  ...  ...              ...              ...   \n",
       "19821        2260.813993  ...      7358.802631     10632.870032   \n",
       "19822        1403.000191  ...      9174.082484      8283.044866   \n",
       "19823         306.821038  ...        47.994576        67.531582   \n",
       "19824          47.496680  ...       283.767930        90.745563   \n",
       "19825         269.647255  ...       554.444802       511.894292   \n",
       "\n",
       "       Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  Vic_1 (GE) - TPM  \\\n",
       "0             8.392188         0.000000         0.000000          0.000000   \n",
       "1           385.693391       383.276940       406.241639        276.773111   \n",
       "2           260.216487       177.158771       215.282147        162.996568   \n",
       "3           327.976622       676.676655       632.319698        622.953497   \n",
       "4           128.419995       108.867940       232.579308        223.051153   \n",
       "...                ...              ...              ...               ...   \n",
       "19821     12801.790715     10104.657626     10909.565036       9206.009642   \n",
       "19822      4301.461428      4804.391257      3884.423218       3057.111744   \n",
       "19823       250.489120       247.530226       322.815816        209.147013   \n",
       "19824       208.279627       163.870817       182.534512        105.147091   \n",
       "19825       865.093206       878.869355       807.683394        691.603197   \n",
       "\n",
       "       Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \\\n",
       "0              0.000000          0.000000      Vibrio   \n",
       "1            178.371963        223.310216      Vibrio   \n",
       "2             97.415072        128.631521      Vibrio   \n",
       "3            648.744407        603.585147      Vibrio   \n",
       "4            367.948661        124.616224      Vibrio   \n",
       "...                 ...               ...         ...   \n",
       "19821       6822.881422       4465.778497       KLEBS   \n",
       "19822       3325.739216       3063.158487       KLEBS   \n",
       "19823        255.838672        283.460097       KLEBS   \n",
       "19824        141.333099        141.038356       KLEBS   \n",
       "19825        745.262717        564.096144       KLEBS   \n",
       "\n",
       "                                             upstream200  \n",
       "0      CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...  \n",
       "1      ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...  \n",
       "2      TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...  \n",
       "3      TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...  \n",
       "4      CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...  \n",
       "...                                                  ...  \n",
       "19821  GTAAATCAATTTGTTATCGCCACTTAGTTAAAAAAAATTTTTTTTT...  \n",
       "19822  AATGGCGCTGCTTTGCGTATCCCGCAGCCGTTCATGAAGCAGCTTG...  \n",
       "19823  AAGCTTCTCCAGCAGGAGCAGAGACTGATGCTGGATGAATTTAGCC...  \n",
       "19824  GATGCTTGTCAGGGGGGCGGAGCCTATGGAAAAACGGCCGTTGTGC...  \n",
       "19825  AGGCTTCTTATGACGAAACAACAGAAAACAGCGCTGAACATGGCTA...  \n",
       "\n",
       "[19826 rows x 41 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f46254e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_column_name(col_name):\n",
    "    \"\"\"\n",
    "    clean and format column name by replacing spaces and special characters with underscores and converting to lowercase.\n",
    "\n",
    "    Parameters:\n",
    "    col_name (str): The original column name.\n",
    "\n",
    "    Returns:\n",
    "    str: The cleaned and formatted column name.\n",
    "    \"\"\"\n",
    "    col_name = col_name.replace(\"_\", \" \")\n",
    "    col_name = col_name.replace(\"-\", \" \")\n",
    "    col_name = re.sub(r\"[^\\w\\s]\", \" \", col_name)\n",
    "    col_name = re.sub(r\"\\s+\", \" \", col_name)\n",
    "    # Replace non-word characters (except for spaces) with nothing\n",
    "    col_name = re.sub(r\"[^\\w\\s]\", \"\", col_name)\n",
    "    # Replace spaces with underscores\n",
    "    col_name = col_name.replace(\" \", \"_\")\n",
    "    # Convert to lowercase\n",
    "    cleaned_name = col_name.lower()\n",
    "    return cleaned_name\n",
    "\n",
    "\n",
    "def rename_columns(df):\n",
    "    \"\"\"\n",
    "    Rename all columns of the DataFrame to a more convenient format.\n",
    "\n",
    "    Parameters:\n",
    "    df (pd.DataFrame): The DataFrame whose columns are to be renamed.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: DataFrame with renamed columns.\n",
    "    \"\"\"\n",
    "    # Create a dictionary to map old column names to new column names\n",
    "    new_columns = {col: clean_column_name(col) for col in df.columns}\n",
    "\n",
    "    # Rename columns in the DataFrame\n",
    "    df.rename(columns=new_columns, inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dee6e3df",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = rename_columns(merged_df)\n",
    "\n",
    "# Extract and clean regions\n",
    "merged_df[\"is_complement\"] = merged_df[\"region\"].str.contains(\"complement\")\n",
    "merged_df[\"region_clean\"] = merged_df[\"region\"].str.replace(\n",
    "    \"complement\\(|\\)\", \"\", regex=True\n",
    ")\n",
    "merged_df[\"is_circular\"] = merged_df[\"region_clean\"].str.contains(\"join\")\n",
    "merged_df[\"region_clean\"] = merged_df[\"region_clean\"].str.replace(\n",
    "    \"join\\(|\\)\", \"\", regex=True\n",
    ")\n",
    "\n",
    "\n",
    "def parse_region(region):\n",
    "    start, *intermediate, end = region.split(\"..\")\n",
    "    if not intermediate:\n",
    "        end = int(end)\n",
    "        start = int(start)\n",
    "        length = end - start + 1\n",
    "        return start, end, length\n",
    "    else:\n",
    "        first_end, second_start = intermediate[0].split(\",\")\n",
    "        length1 = int(first_end) - int(start) + 1\n",
    "        length2 = int(end) - int(second_start) + 1\n",
    "        return int(start), int(end), length1 + length2\n",
    "\n",
    "\n",
    "merged_df[[\"region_start\", \"region_end\", \"region_length\"]] = merged_df[\n",
    "    \"region_clean\"\n",
    "].apply(lambda x: pd.Series(parse_region(x)))\n",
    "merged_df.drop(columns=[\"region\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "28977637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>as_1_ge_tpm</th>\n",
       "      <th>as_2_ge_tpm</th>\n",
       "      <th>as_3_ge_tpm</th>\n",
       "      <th>bs_1_ge_tpm</th>\n",
       "      <th>bs_2_ge_tpm</th>\n",
       "      <th>bs_3_ge_tpm</th>\n",
       "      <th>ctrl_1_ge_tpm</th>\n",
       "      <th>ctrl_2_ge_tpm</th>\n",
       "      <th>...</th>\n",
       "      <th>vic_2_ge_tpm</th>\n",
       "      <th>vic_3_ge_tpm</th>\n",
       "      <th>species_id</th>\n",
       "      <th>upstream200</th>\n",
       "      <th>is_complement</th>\n",
       "      <th>region_clean</th>\n",
       "      <th>is_circular</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>region_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...</td>\n",
       "      <td>True</td>\n",
       "      <td>235..402</td>\n",
       "      <td>False</td>\n",
       "      <td>235</td>\n",
       "      <td>402</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>393.166189</td>\n",
       "      <td>311.937046</td>\n",
       "      <td>383.829258</td>\n",
       "      <td>377.730614</td>\n",
       "      <td>273.496217</td>\n",
       "      <td>331.362458</td>\n",
       "      <td>315.921082</td>\n",
       "      <td>342.779023</td>\n",
       "      <td>...</td>\n",
       "      <td>178.371963</td>\n",
       "      <td>223.310216</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...</td>\n",
       "      <td>True</td>\n",
       "      <td>372..806</td>\n",
       "      <td>False</td>\n",
       "      <td>372</td>\n",
       "      <td>806</td>\n",
       "      <td>435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>162.400143</td>\n",
       "      <td>171.281001</td>\n",
       "      <td>174.959500</td>\n",
       "      <td>261.605512</td>\n",
       "      <td>228.250776</td>\n",
       "      <td>233.104257</td>\n",
       "      <td>154.689875</td>\n",
       "      <td>184.843102</td>\n",
       "      <td>...</td>\n",
       "      <td>97.415072</td>\n",
       "      <td>128.631521</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...</td>\n",
       "      <td>True</td>\n",
       "      <td>816..2210</td>\n",
       "      <td>False</td>\n",
       "      <td>816</td>\n",
       "      <td>2210</td>\n",
       "      <td>1395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>720.122251</td>\n",
       "      <td>825.256114</td>\n",
       "      <td>860.384381</td>\n",
       "      <td>560.678176</td>\n",
       "      <td>683.429275</td>\n",
       "      <td>731.909515</td>\n",
       "      <td>554.730307</td>\n",
       "      <td>697.202002</td>\n",
       "      <td>...</td>\n",
       "      <td>648.744407</td>\n",
       "      <td>603.585147</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...</td>\n",
       "      <td>True</td>\n",
       "      <td>2271..3896</td>\n",
       "      <td>False</td>\n",
       "      <td>2271</td>\n",
       "      <td>3896</td>\n",
       "      <td>1626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>NC_002505</td>\n",
       "      <td>132.002479</td>\n",
       "      <td>371.588303</td>\n",
       "      <td>244.840701</td>\n",
       "      <td>150.111163</td>\n",
       "      <td>232.967463</td>\n",
       "      <td>208.688639</td>\n",
       "      <td>205.699493</td>\n",
       "      <td>245.932501</td>\n",
       "      <td>...</td>\n",
       "      <td>367.948661</td>\n",
       "      <td>124.616224</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...</td>\n",
       "      <td>True</td>\n",
       "      <td>3899..4156</td>\n",
       "      <td>False</td>\n",
       "      <td>3899</td>\n",
       "      <td>4156</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19821</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>2196.775545</td>\n",
       "      <td>2257.918323</td>\n",
       "      <td>2155.878280</td>\n",
       "      <td>1635.286545</td>\n",
       "      <td>2357.383900</td>\n",
       "      <td>1929.630652</td>\n",
       "      <td>2260.813993</td>\n",
       "      <td>2388.506147</td>\n",
       "      <td>...</td>\n",
       "      <td>6822.881422</td>\n",
       "      <td>4465.778497</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>GTAAATCAATTTGTTATCGCCACTTAGTTAAAAAAAATTTTTTTTT...</td>\n",
       "      <td>False</td>\n",
       "      <td>1433..1663</td>\n",
       "      <td>False</td>\n",
       "      <td>1433</td>\n",
       "      <td>1663</td>\n",
       "      <td>231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19822</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>2181.385479</td>\n",
       "      <td>1981.081622</td>\n",
       "      <td>1666.816029</td>\n",
       "      <td>1746.367340</td>\n",
       "      <td>1849.783568</td>\n",
       "      <td>1862.333827</td>\n",
       "      <td>1403.000191</td>\n",
       "      <td>1621.881235</td>\n",
       "      <td>...</td>\n",
       "      <td>3325.739216</td>\n",
       "      <td>3063.158487</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>AATGGCGCTGCTTTGCGTATCCCGCAGCCGTTCATGAAGCAGCTTG...</td>\n",
       "      <td>False</td>\n",
       "      <td>1660..1998</td>\n",
       "      <td>False</td>\n",
       "      <td>1660</td>\n",
       "      <td>1998</td>\n",
       "      <td>339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19823</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>107.262956</td>\n",
       "      <td>165.195837</td>\n",
       "      <td>160.945638</td>\n",
       "      <td>268.944256</td>\n",
       "      <td>336.767008</td>\n",
       "      <td>251.302784</td>\n",
       "      <td>306.821038</td>\n",
       "      <td>372.654817</td>\n",
       "      <td>...</td>\n",
       "      <td>255.838672</td>\n",
       "      <td>283.460097</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>AAGCTTCTCCAGCAGGAGCAGAGACTGATGCTGGATGAATTTAGCC...</td>\n",
       "      <td>True</td>\n",
       "      <td>2287..2544</td>\n",
       "      <td>False</td>\n",
       "      <td>2287</td>\n",
       "      <td>2544</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19824</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>59.660108</td>\n",
       "      <td>97.389121</td>\n",
       "      <td>74.907776</td>\n",
       "      <td>46.934265</td>\n",
       "      <td>44.546468</td>\n",
       "      <td>41.170402</td>\n",
       "      <td>47.496680</td>\n",
       "      <td>83.273274</td>\n",
       "      <td>...</td>\n",
       "      <td>141.333099</td>\n",
       "      <td>141.038356</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>GATGCTTGTCAGGGGGGCGGAGCCTATGGAAAAACGGCCGTTGTGC...</td>\n",
       "      <td>False</td>\n",
       "      <td>2667..2858</td>\n",
       "      <td>False</td>\n",
       "      <td>2667</td>\n",
       "      <td>2858</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19825</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae MGH 78578</td>\n",
       "      <td>NC_009653</td>\n",
       "      <td>464.105365</td>\n",
       "      <td>520.751232</td>\n",
       "      <td>461.618791</td>\n",
       "      <td>264.223870</td>\n",
       "      <td>379.004922</td>\n",
       "      <td>324.158181</td>\n",
       "      <td>269.647255</td>\n",
       "      <td>317.645814</td>\n",
       "      <td>...</td>\n",
       "      <td>745.262717</td>\n",
       "      <td>564.096144</td>\n",
       "      <td>KLEBS</td>\n",
       "      <td>AGGCTTCTTATGACGAAACAACAGAAAACAGCGCTGAACATGGCTA...</td>\n",
       "      <td>False</td>\n",
       "      <td>2858..3259</td>\n",
       "      <td>False</td>\n",
       "      <td>2858</td>\n",
       "      <td>3259</td>\n",
       "      <td>402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19826 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 species chromosome  \\\n",
       "0           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "1           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "2           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "3           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "4           Vibrio cholerae O1 biovar El Tor str. N16961  NC_002505   \n",
       "...                                                  ...        ...   \n",
       "19821  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19822  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19823  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19824  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "19825  Klebsiella pneumoniae subsp. pneumoniae MGH 78578  NC_009653   \n",
       "\n",
       "       as_1_ge_tpm  as_2_ge_tpm  as_3_ge_tpm  bs_1_ge_tpm  bs_2_ge_tpm  \\\n",
       "0         0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "1       393.166189   311.937046   383.829258   377.730614   273.496217   \n",
       "2       162.400143   171.281001   174.959500   261.605512   228.250776   \n",
       "3       720.122251   825.256114   860.384381   560.678176   683.429275   \n",
       "4       132.002479   371.588303   244.840701   150.111163   232.967463   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "19821  2196.775545  2257.918323  2155.878280  1635.286545  2357.383900   \n",
       "19822  2181.385479  1981.081622  1666.816029  1746.367340  1849.783568   \n",
       "19823   107.262956   165.195837   160.945638   268.944256   336.767008   \n",
       "19824    59.660108    97.389121    74.907776    46.934265    44.546468   \n",
       "19825   464.105365   520.751232   461.618791   264.223870   379.004922   \n",
       "\n",
       "       bs_3_ge_tpm  ctrl_1_ge_tpm  ctrl_2_ge_tpm  ...  vic_2_ge_tpm  \\\n",
       "0         0.000000       0.000000       0.000000  ...      0.000000   \n",
       "1       331.362458     315.921082     342.779023  ...    178.371963   \n",
       "2       233.104257     154.689875     184.843102  ...     97.415072   \n",
       "3       731.909515     554.730307     697.202002  ...    648.744407   \n",
       "4       208.688639     205.699493     245.932501  ...    367.948661   \n",
       "...            ...            ...            ...  ...           ...   \n",
       "19821  1929.630652    2260.813993    2388.506147  ...   6822.881422   \n",
       "19822  1862.333827    1403.000191    1621.881235  ...   3325.739216   \n",
       "19823   251.302784     306.821038     372.654817  ...    255.838672   \n",
       "19824    41.170402      47.496680      83.273274  ...    141.333099   \n",
       "19825   324.158181     269.647255     317.645814  ...    745.262717   \n",
       "\n",
       "       vic_3_ge_tpm  species_id  \\\n",
       "0          0.000000      Vibrio   \n",
       "1        223.310216      Vibrio   \n",
       "2        128.631521      Vibrio   \n",
       "3        603.585147      Vibrio   \n",
       "4        124.616224      Vibrio   \n",
       "...             ...         ...   \n",
       "19821   4465.778497       KLEBS   \n",
       "19822   3063.158487       KLEBS   \n",
       "19823    283.460097       KLEBS   \n",
       "19824    141.038356       KLEBS   \n",
       "19825    564.096144       KLEBS   \n",
       "\n",
       "                                             upstream200  is_complement  \\\n",
       "0      CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...           True   \n",
       "1      ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...           True   \n",
       "2      TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...           True   \n",
       "3      TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...           True   \n",
       "4      CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...           True   \n",
       "...                                                  ...            ...   \n",
       "19821  GTAAATCAATTTGTTATCGCCACTTAGTTAAAAAAAATTTTTTTTT...          False   \n",
       "19822  AATGGCGCTGCTTTGCGTATCCCGCAGCCGTTCATGAAGCAGCTTG...          False   \n",
       "19823  AAGCTTCTCCAGCAGGAGCAGAGACTGATGCTGGATGAATTTAGCC...           True   \n",
       "19824  GATGCTTGTCAGGGGGGCGGAGCCTATGGAAAAACGGCCGTTGTGC...          False   \n",
       "19825  AGGCTTCTTATGACGAAACAACAGAAAACAGCGCTGAACATGGCTA...          False   \n",
       "\n",
       "       region_clean  is_circular  region_start  region_end  region_length  \n",
       "0          235..402        False           235         402            168  \n",
       "1          372..806        False           372         806            435  \n",
       "2         816..2210        False           816        2210           1395  \n",
       "3        2271..3896        False          2271        3896           1626  \n",
       "4        3899..4156        False          3899        4156            258  \n",
       "...             ...          ...           ...         ...            ...  \n",
       "19821    1433..1663        False          1433        1663            231  \n",
       "19822    1660..1998        False          1660        1998            339  \n",
       "19823    2287..2544        False          2287        2544            258  \n",
       "19824    2667..2858        False          2667        2858            192  \n",
       "19825    2858..3259        False          2858        3259            402  \n",
       "\n",
       "[19826 rows x 46 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8baa502a-5e70-4a5b-886d-07e81d586306",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv(f\"{os.getcwd()}/data/combined_data.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
