{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d43146eb-dd1e-44dd-af6c-1ce66157518c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7a7699a-bfa1-4b61-890b-5f28448fb536",
   "metadata": {},
   "source": [
    "# Expression data section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "48f18357-f216-4ff9-a512-29b3abc78b3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_species_id_and_column_name(column_name, reference_columns):\n",
    "    parts = column_name.split('_')\n",
    "    if '_'.join(parts[2:]) in reference_columns:\n",
    "        return '_'.join(parts[:2]), '_'.join(parts[2:])\n",
    "    else:\n",
    "        return parts[0], '_'.join(parts[1:])\n",
    "\n",
    "def read_xlsx_files_to_dataframe(directory, concatenated_df=None):\n",
    "    # Initialize an empty dataframe if concatenated_df is not provided\n",
    "    if concatenated_df is None:\n",
    "        concatenated_df = pd.DataFrame()\n",
    "    \n",
    "    # Get the set of species that have already been read\n",
    "    existing_species = set(concatenated_df['Species']) if 'Species' in concatenated_df.columns else set()\n",
    "    \n",
    "    # Reference columns for species ID extraction\n",
    "    reference_columns = {col.split('_')[0] for col in concatenated_df.columns if 'TPM' in col}\n",
    "    reference_columns.update({'_'.join(col.split('_')[:2]) for col in concatenated_df.columns if 'TPM' in col})\n",
    "    \n",
    "    # Iterate over all files in the given directory\n",
    "    for filename in os.listdir(directory):\n",
    "        if filename.endswith(\".xlsx\"):\n",
    "            file_path = os.path.join(directory, filename)\n",
    "            \n",
    "            # Remove the extension from the filename\n",
    "            base_filename = os.path.splitext(filename)[0]\n",
    "            \n",
    "            # Read the Excel file into a pandas dataframe\n",
    "            df = pd.read_excel(file_path)\n",
    "            \n",
    "            # Check if the species in the file is already read\n",
    "            if 'Species' in df.columns:\n",
    "                species_in_file = set(df['Species'])\n",
    "                if species_in_file.intersection(existing_species):\n",
    "                    print(f\"Species {species_in_file} is skipped as it has already been read from {base_filename}.\")\n",
    "                    continue\n",
    "            \n",
    "            # Extract 'Species' and 'Region' columns\n",
    "            columns_to_extract = ['Species', 'Chromosome', 'Region']\n",
    "            \n",
    "            # Identify TPM columns and extract them\n",
    "            tpm_columns = [col for col in df.columns if 'TPM' in col]\n",
    "            \n",
    "            # Create a new dataframe with the required columns\n",
    "            extracted_data = df[columns_to_extract + tpm_columns].copy()\n",
    "            \n",
    "            # Extract species id from the TPM columns and rename them\n",
    "            species_ids = []\n",
    "            for col in tpm_columns:\n",
    "                species_id, new_col_name = extract_species_id_and_column_name(col, reference_columns)\n",
    "                extracted_data[new_col_name] = extracted_data[col]\n",
    "                extracted_data.drop(columns=[col], inplace=True)\n",
    "                species_ids.append(species_id)\n",
    "            \n",
    "            try:\n",
    "                if len(set(species_ids)) == 1:\n",
    "                    extracted_data['Species ID'] = species_ids[0]\n",
    "                else:\n",
    "                    raise ValueError(\"Multiple species IDs found in one file.\")\n",
    "            except ValueError as e:\n",
    "                print(f\"Error processing file {filename}: {e}\")\n",
    "                extracted_data['Species ID'] = species_ids[0]\n",
    "\n",
    "            # Concatenate the extracted data to the main dataframe\n",
    "            if concatenated_df.empty:\n",
    "                concatenated_df = extracted_data\n",
    "            else:\n",
    "                if set(concatenated_df.columns) == set(extracted_data.columns):\n",
    "                    concatenated_df = pd.concat([concatenated_df, extracted_data], ignore_index=True)\n",
    "                    print(f\"Data from {filename} is extracted\")\n",
    "\n",
    "                else:\n",
    "                    unmatched_columns = set(concatenated_df.columns).symmetric_difference(set(extracted_data.columns))\n",
    "                    print(f\"Column names do not match across the files. Unmatched columns: {unmatched_columns}\")\n",
    "                    return concatenated_df\n",
    "    \n",
    "    return concatenated_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ed5acfd3-9a4f-4a6c-a367-a931301c78f0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACHX As_1 (GE) - TPM\n",
      "ACHX As_2 (GE) - TPM\n",
      "ACHX As_3 (GE) - TPM\n",
      "ACHX Bs_1 (GE) - TPM\n",
      "ACHX Bs_2 (GE) - TPM\n",
      "ACHX Bs_3 (GE) - TPM\n",
      "ACHX Ctrl_1 (GE) - TPM\n",
      "ACHX Ctrl_2 (GE) - TPM\n",
      "ACHX Ctrl_3 (GE) - TPM\n",
      "ACHX Li_1 (GE) - TPM\n",
      "ACHX Li_2 (GE) - TPM\n",
      "ACHX Li_3 (GE) - TPM\n",
      "ACHX Mig_1 (GE) - TPM\n",
      "ACHX Mig_2 (GE) - TPM\n",
      "ACHX Mig_3 (GE) - TPM\n",
      "ACHX Nd_1 (GE) - TPM\n",
      "ACHX Nd_2 (GE) - TPM\n",
      "ACHX Nd_3 (GE) - TPM\n",
      "ACHX Ns_1 (GE) - TPM\n",
      "ACHX Ns_2 (GE) - TPM\n",
      "ACHX Ns_3 (GE) - TPM\n",
      "ACHX Oss_1 (GE) - TPM\n",
      "ACHX Oss_2 (GE) - TPM\n",
      "ACHX Oss_3 (GE) - TPM\n",
      "ACHX Oxs_1 (GE) - TPM\n",
      "ACHX Oxs_2 (GE) - TPM\n",
      "ACHX Oxs_3 (GE) - TPM\n",
      "ACHX Sp_1 (GE) - TPM\n",
      "ACHX Sp_2 (GE) - TPM\n",
      "ACHX Sp_3 (GE) - TPM\n",
      "ACHX Tm_1 (GE) - TPM\n",
      "ACHX Tm_2 (GE) - TPM\n",
      "ACHX Tm_3 (GE) - TPM\n",
      "ACHX Vic_1 (GE) - TPM\n",
      "ACHX Vic_2 (GE) - TPM\n",
      "ACHX Vic_3 (GE) - TPM\n",
      "ACIB As_1 (GE) - TPM\n",
      "ACIB As_2 (GE) - TPM\n",
      "ACIB As_3 (GE) - TPM\n",
      "ACIB Bs_1 (GE) - TPM\n",
      "ACIB Bs_2 (GE) - TPM\n",
      "ACIB Bs_3 (GE) - TPM\n",
      "ACIB Ctrl_1 (GE) - TPM\n",
      "ACIB Ctrl_2 (GE) - TPM\n",
      "ACIB Ctrl_3 (GE) - TPM\n",
      "ACIB Li_1 (GE) - TPM\n",
      "ACIB Li_2 (GE) - TPM\n",
      "ACIB Li_3 (GE) - TPM\n",
      "ACIB Mig_1 (GE) - TPM\n",
      "ACIB Mig_2 (GE) - TPM\n",
      "ACIB Mig_3 (GE) - TPM\n",
      "ACIB Nd_1 (GE) - TPM\n",
      "ACIB Nd_2 (GE) - TPM\n",
      "ACIB Nd_3 (GE) - TPM\n",
      "ACIB Ns_1 (GE) - TPM\n",
      "ACIB Ns_2 (GE) - TPM\n",
      "ACIB Ns_3 (GE) - TPM\n",
      "ACIB Oss_1 (GE) - TPM\n",
      "ACIB Oss_2 (GE) - TPM\n",
      "ACIB Oss_3 (GE) - TPM\n",
      "ACIB Oxs_1 (GE) - TPM\n",
      "ACIB Oxs_2 (GE) - TPM\n",
      "ACIB Oxs_3 (GE) - TPM\n",
      "ACIB Sp_1 (GE) - TPM\n",
      "ACIB Sp_2 (GE) - TPM\n",
      "ACIB Sp_3 (GE) - TPM\n",
      "ACIB Tm_1 (GE) - TPM\n",
      "ACIB Tm_2 (GE) - TPM\n",
      "ACIB Tm_3 (GE) - TPM\n",
      "ACIB Vic_1 (GE) - TPM\n",
      "ACIB Vic_2 (GE) - TPM\n",
      "ACIB Vic_3 (GE) - TPM\n",
      "AGGA As_1 (GE) - TPM\n",
      "AGGA As_2 (GE) - TPM\n",
      "AGGA As_3 (GE) - TPM\n",
      "AGGA Bs_1 (GE) - TPM\n",
      "AGGA Bs_2 (GE) - TPM\n",
      "AGGA Bs_3 (GE) - TPM\n",
      "AGGA Ctrl_1 (GE) - TPM\n",
      "AGGA Ctrl_2 (GE) - TPM\n",
      "AGGA Ctrl_3 (GE) - TPM\n",
      "AGGA Li_1 (GE) - TPM\n",
      "AGGA Li_2 (GE) - TPM\n",
      "AGGA Li_3 (GE) - TPM\n",
      "AGGA Mig_1 (GE) - TPM\n",
      "AGGA Mig_2 (GE) - TPM\n",
      "AGGA Mig_3 (GE) - TPM\n",
      "AGGA Nd_1 (GE) - TPM\n",
      "AGGA Nd_2 (GE) - TPM\n",
      "AGGA Nd_3 (GE) - TPM\n",
      "AGGA Ns_1 (GE) - TPM\n",
      "AGGA Ns_2 (GE) - TPM\n",
      "AGGA Ns_3 (GE) - TPM\n",
      "AGGA Oss_1 (GE) - TPM\n",
      "AGGA Oss_2 (GE) - TPM\n",
      "AGGA Oss_3 (GE) - TPM\n",
      "AGGA Oxs_1 (GE) - TPM\n",
      "AGGA Oxs_2 (GE) - TPM\n",
      "AGGA Oxs_3 (GE) - TPM\n",
      "AGGA Sp_1 (GE) - TPM\n",
      "AGGA Sp_2 (GE) - TPM\n",
      "AGGA Sp_3 (GE) - TPM\n",
      "AGGA Tm_1 (GE) - TPM\n",
      "AGGA Tm_2 (GE) - TPM\n",
      "AGGA Tm_3 (GE) - TPM\n",
      "AGGA Vic_1 (GE) - TPM\n",
      "AGGA Vic_2 (GE) - TPM\n",
      "AGGA Vic_3 (GE) - TPM\n",
      "BBURG As_1 (GE) - TPM\n",
      "BBURG As_2 (GE) - TPM\n",
      "BBURG As_3 (GE) - TPM\n",
      "BBURG Bs_1 (GE) - TPM\n",
      "BBURG Bs_2 (GE) - TPM\n",
      "BBURG Bs_3 (GE) - TPM\n",
      "BBURG Ctrl_1 (GE) - TPM\n",
      "BBURG Ctrl_2 (GE) - TPM\n",
      "BBURG Ctrl_3 (GE) - TPM\n",
      "BBURG Li_1 (GE) - TPM\n",
      "BBURG Li_2 (GE) - TPM\n",
      "BBURG Li_3 (GE) - TPM\n",
      "BBURG Mig_1 (GE) - TPM\n",
      "BBURG Mig_2 (GE) - TPM\n",
      "BBURG Mig_3 (GE) - TPM\n",
      "BBURG Nd_1 (GE) - TPM\n",
      "BBURG Nd_2 (GE) - TPM\n",
      "BBURG Nd_3 (GE) - TPM\n",
      "BBURG Ns_1 (GE) - TPM\n",
      "BBURG Ns_2 (GE) - TPM\n",
      "BBURG Ns_3 (GE) - TPM\n",
      "BBURG Oss_1 (GE) - TPM\n",
      "BBURG Oss_2 (GE) - TPM\n",
      "BBURG Oss_3 (GE) - TPM\n",
      "BBURG Oxs_1 (GE) - TPM\n",
      "BBURG Oxs_2 (GE) - TPM\n",
      "BBURG Oxs_3 (GE) - TPM\n",
      "BBURG Sp_1 (GE) - TPM\n",
      "BBURG Sp_2 (GE) - TPM\n",
      "BBURG Sp_3 (GE) - TPM\n",
      "BBURG Tm_1 (GE) - TPM\n",
      "BBURG Tm_2 (GE) - TPM\n",
      "BBURG Tm_3 (GE) - TPM\n",
      "BBURG Vic_1 (GE) - TPM\n",
      "BBURG Vic_2 (GE) - TPM\n",
      "BBURG Vic_3 (GE) - TPM\n",
      "BURK As_1 (GE) - TPM\n",
      "BURK As_2 (GE) - TPM\n",
      "BURK As_3 (GE) - TPM\n",
      "BURK Bs_1 (GE) - TPM\n",
      "BURK Bs_2 (GE) - TPM\n",
      "BURK Bs_3 (GE) - TPM\n",
      "BURK Ctrl_1 (GE) - TPM\n",
      "BURK Ctrl_2 (GE) - TPM\n",
      "BURK Ctrl_3 (GE) - TPM\n",
      "BURK Li_1 (GE) - TPM\n",
      "BURK Li_2 (GE) - TPM\n",
      "BURK Li_3 (GE) - TPM\n",
      "BURK Mig_1 (GE) - TPM\n",
      "BURK Mig_2 (GE) - TPM\n",
      "BURK Mig_3 (GE) - TPM\n",
      "BURK Nd_1 (GE) - TPM\n",
      "BURK Nd_2 (GE) - TPM\n",
      "BURK Nd_3 (GE) - TPM\n",
      "BURK Ns_1 (GE) - TPM\n",
      "BURK Ns_2 (GE) - TPM\n",
      "BURK Ns_3 (GE) - TPM\n",
      "BURK Oss_1 (GE) - TPM\n",
      "BURK Oss_2 (GE) - TPM\n",
      "BURK Oss_3 (GE) - TPM\n",
      "BURK Oxs_1 (GE) - TPM\n",
      "BURK Oxs_2 (GE) - TPM\n",
      "BURK Oxs_3 (GE) - TPM\n",
      "BURK Sp_1 (GE) - TPM\n",
      "BURK Sp_2 (GE) - TPM\n",
      "BURK Sp_3 (GE) - TPM\n",
      "BURK Tm_1 (GE) - TPM\n",
      "BURK Tm_2 (GE) - TPM\n",
      "BURK Tm_3 (GE) - TPM\n",
      "BURK Vic_1 (GE) - TPM\n",
      "BURK Vic_2 (GE) - TPM\n",
      "BURK Vic_3 (GE) - TPM\n",
      "Campy As_1 (GE) - TPM\n",
      "Campy As_2 (GE) - TPM\n",
      "Campy As_3 (GE) - TPM\n",
      "Campy Bs_1 (GE) - TPM\n",
      "Campy Bs_2 (GE) - TPM\n",
      "Campy Bs_3 (GE) - TPM\n",
      "Campy Ctrl_1 (GE) - TPM\n",
      "Campy Ctrl_2 (GE) - TPM\n",
      "Campy Ctrl_3 (GE) - TPM\n",
      "Campy Li_1 (GE) - TPM\n",
      "Campy Li_2 (GE) - TPM\n",
      "Campy Li_3 (GE) - TPM\n",
      "Campy Mig_1 (GE) - TPM\n",
      "Campy Mig_2 (GE) - TPM\n",
      "Campy Mig_3 (GE) - TPM\n",
      "Campy Nd_1 (GE) - TPM\n",
      "Campy Nd_2 (GE) - TPM\n",
      "Campy Nd_3 (GE) - TPM\n",
      "Campy Ns_1 (GE) - TPM\n",
      "Campy Ns_2 (GE) - TPM\n",
      "Campy Ns_3 (GE) - TPM\n",
      "Campy Oss_1 (GE) - TPM\n",
      "Campy Oss_2 (GE) - TPM\n",
      "Campy Oss_3 (GE) - TPM\n",
      "Campy Oxs_1 (GE) - TPM\n",
      "Campy Oxs_2 (GE) - TPM\n",
      "Campy Oxs_3 (GE) - TPM\n",
      "Campy Sp_1 (GE) - TPM\n",
      "Campy Sp_2 (GE) - TPM\n",
      "Campy Sp_3 (GE) - TPM\n",
      "Campy Tm_1 (GE) - TPM\n",
      "Campy Tm_2 (GE) - TPM\n",
      "Campy Tm_3 (GE) - TPM\n",
      "Campy Vic_1 (GE) - TPM\n",
      "Campy Vic_2 (GE) - TPM\n",
      "Campy Vic_3 (GE) - TPM\n",
      "ENTFA As_1 (GE) - TPM\n",
      "ENTFA As_2 (GE) - TPM\n",
      "ENTFA As_3 (GE) - TPM\n",
      "ENTFA Bs_1 (GE) - TPM\n",
      "ENTFA Bs_2 (GE) - TPM\n",
      "ENTFA Bs_3 (GE) - TPM\n",
      "ENTFA Ctrl_1 (GE) - TPM\n",
      "ENTFA Ctrl_2 (GE) - TPM\n",
      "ENTFA Ctrl_3 (GE) - TPM\n",
      "ENTFA Li_1 (GE) - TPM\n",
      "ENTFA Li_2 (GE) - TPM\n",
      "ENTFA Li_3 (GE) - TPM\n",
      "ENTFA Mig_1 (GE) - TPM\n",
      "ENTFA Mig_2 (GE) - TPM\n",
      "ENTFA Mig_3 (GE) - TPM\n",
      "ENTFA Nd_1 (GE) - TPM\n",
      "ENTFA Nd_2 (GE) - TPM\n",
      "ENTFA Nd_3 (GE) - TPM\n",
      "ENTFA Ns_1 (GE) - TPM\n",
      "ENTFA Ns_2 (GE) - TPM\n",
      "ENTFA Ns_3 (GE) - TPM\n",
      "ENTFA Oss_1 (GE) - TPM\n",
      "ENTFA Oss_2 (GE) - TPM\n",
      "ENTFA Oss_3 (GE) - TPM\n",
      "ENTFA Oxs_1 (GE) - TPM\n",
      "ENTFA Oxs_2 (GE) - TPM\n",
      "ENTFA Oxs_3 (GE) - TPM\n",
      "ENTFA Sp_1 (GE) - TPM\n",
      "ENTFA Sp_2 (GE) - TPM\n",
      "ENTFA Sp_3 (GE) - TPM\n",
      "ENTFA Tm_1 (GE) - TPM\n",
      "ENTFA Tm_2 (GE) - TPM\n",
      "ENTFA Tm_3 (GE) - TPM\n",
      "ENTFA Vic_1 (GE) - TPM\n",
      "ENTFA Vic_2 (GE) - TPM\n",
      "ENTFA Vic_3 (GE) - TPM\n",
      "EPEC As_1 (GE) - TPM\n",
      "EPEC As_2 (GE) - TPM\n",
      "EPEC As_3 (GE) - TPM\n",
      "EPEC Bs_1 (GE) - TPM\n",
      "EPEC Bs_2 (GE) - TPM\n",
      "EPEC Bs_3 (GE) - TPM\n",
      "EPEC Ctrl_1 (GE) - TPM\n",
      "EPEC Ctrl_2 (GE) - TPM\n",
      "EPEC Ctrl_3 (GE) - TPM\n",
      "EPEC Li_1 (GE) - TPM\n",
      "EPEC Li_2 (GE) - TPM\n",
      "EPEC Li_3 (GE) - TPM\n",
      "EPEC Mig_1 (GE) - TPM\n",
      "EPEC Mig_2 (GE) - TPM\n",
      "EPEC Mig_3 (GE) - TPM\n",
      "EPEC Nd_1 (GE) - TPM\n",
      "EPEC Nd_2 (GE) - TPM\n",
      "EPEC Nd_3 (GE) - TPM\n",
      "EPEC Ns_1 (GE) - TPM\n",
      "EPEC Ns_2 (GE) - TPM\n",
      "EPEC Ns_3 (GE) - TPM\n",
      "EPEC Oss_1 (GE) - TPM\n",
      "EPEC Oss_2 (GE) - TPM\n",
      "EPEC Oss_3 (GE) - TPM\n",
      "EPEC Oxs_1 (GE) - TPM\n",
      "EPEC Oxs_2 (GE) - TPM\n",
      "EPEC Oxs_3 (GE) - TPM\n",
      "EPEC Sp_1 (GE) - TPM\n",
      "EPEC Sp_2 (GE) - TPM\n",
      "EPEC Sp_3 (GE) - TPM\n",
      "EPEC Tm_1 (GE) - TPM\n",
      "EPEC Tm_2 (GE) - TPM\n",
      "EPEC Tm_3 (GE) - TPM\n",
      "EPEC Vic_1 (GE) - TPM\n",
      "EPEC Vic_2 (GE) - TPM\n",
      "EPEC Vic_3 (GE) - TPM\n",
      "ETEC As_1 (GE) - TPM\n",
      "ETEC As_2 (GE) - TPM\n",
      "ETEC As_3 (GE) - TPM\n",
      "ETEC Bs_1 (GE) - TPM\n",
      "ETEC Bs_2 (GE) - TPM\n",
      "ETEC Bs_3 (GE) - TPM\n",
      "ETEC Ctrl_1 (GE) - TPM\n",
      "ETEC Ctrl_2 (GE) - TPM\n",
      "ETEC Ctrl_3 (GE) - TPM\n",
      "ETEC Li_1 (GE) - TPM\n",
      "ETEC Li_2 (GE) - TPM\n",
      "ETEC Li_3 (GE) - TPM\n",
      "ETEC Mig_1 (GE) - TPM\n",
      "ETEC Mig_2 (GE) - TPM\n",
      "ETEC Mig_3 (GE) - TPM\n",
      "ETEC Nd_1 (GE) - TPM\n",
      "ETEC Nd_2 (GE) - TPM\n",
      "ETEC Nd_3 (GE) - TPM\n",
      "ETEC Ns_1 (GE) - TPM\n",
      "ETEC Ns_2 (GE) - TPM\n",
      "ETEC Ns_3 (GE) - TPM\n",
      "ETEC Oss_1 (GE) - TPM\n",
      "ETEC Oss_2 (GE) - TPM\n",
      "ETEC Oss_3 (GE) - TPM\n",
      "ETEC Oxs_1 (GE) - TPM\n",
      "ETEC Oxs_2 (GE) - TPM\n",
      "ETEC Oxs_3 (GE) - TPM\n",
      "ETEC Sp_1 (GE) - TPM\n",
      "ETEC Sp_2 (GE) - TPM\n",
      "ETEC Sp_3 (GE) - TPM\n",
      "ETEC Tm_1 (GE) - TPM\n",
      "ETEC Tm_2 (GE) - TPM\n",
      "ETEC Tm_3 (GE) - TPM\n",
      "ETEC Vic_1 (GE) - TPM\n",
      "ETEC Vic_2 (GE) - TPM\n",
      "ETEC Vic_3 (GE) - TPM\n",
      "UPEC As_1 (GE) - TPM\n",
      "UPEC As_2 (GE) - TPM\n",
      "UPEC As_3 (GE) - TPM\n",
      "UPEC Bs_1 (GE) - TPM\n",
      "UPEC Bs_2 (GE) - TPM\n",
      "UPEC Bs_3 (GE) - TPM\n",
      "UPEC Ctrl_1 (GE) - TPM\n",
      "UPEC Ctrl_2 (GE) - TPM\n",
      "UPEC Ctrl_3 (GE) - TPM\n",
      "UPEC Li_1 (GE) - TPM\n",
      "UPEC Li_2 (GE) - TPM\n",
      "UPEC Li_3 (GE) - TPM\n",
      "UPEC Mig_1 (GE) - TPM\n",
      "UPEC Mig_2 (GE) - TPM\n",
      "UPEC Mig_3 (GE) - TPM\n",
      "UPEC Nd_1 (GE) - TPM\n",
      "UPEC Nd_2 (GE) - TPM\n",
      "UPEC Nd_3 (GE) - TPM\n",
      "UPEC Ns_1 (GE) - TPM\n",
      "UPEC Ns_2 (GE) - TPM\n",
      "UPEC Ns_3 (GE) - TPM\n",
      "UPEC Oss_1 (GE) - TPM\n",
      "UPEC Oss_2 (GE) - TPM\n",
      "UPEC Oss_3 (GE) - TPM\n",
      "UPEC Oxs_1 (GE) - TPM\n",
      "UPEC Oxs_2 (GE) - TPM\n",
      "UPEC Oxs_3 (GE) - TPM\n",
      "UPEC Sp_1 (GE) - TPM\n",
      "UPEC Sp_2 (GE) - TPM\n",
      "UPEC Sp_3 (GE) - TPM\n",
      "UPEC Tm_1 (GE) - TPM\n",
      "UPEC Tm_2 (GE) - TPM\n",
      "UPEC Tm_3 (GE) - TPM\n",
      "UPEC Vic_1 (GE) - TPM\n",
      "UPEC Vic_2 (GE) - TPM\n",
      "UPEC Vic_3 (GE) - TPM\n",
      "FRAT As_1 (GE) - TPM\n",
      "FRAT As_2 (GE) - TPM\n",
      "FRAT As_3 (GE) - TPM\n",
      "FRAT Bs_1 (GE) - TPM\n",
      "FRAT Bs_2 (GE) - TPM\n",
      "FRAT Bs_3 (GE) - TPM\n",
      "FRAT Ctrl_1 (GE) - TPM\n",
      "FRAT Ctrl_2 (GE) - TPM\n",
      "FRAT Ctrl_3 (GE) - TPM\n",
      "FRAT Li_1 (GE) - TPM\n",
      "FRAT Li_2 (GE) - TPM\n",
      "FRAT Li_3 (GE) - TPM\n",
      "FRAT Mig_1 (GE) - TPM\n",
      "FRAT Mig_2 (GE) - TPM\n",
      "FRAT Mig_3 (GE) - TPM\n",
      "FRAT Nd_1 (GE) - TPM\n",
      "FRAT Nd_2 (GE) - TPM\n",
      "FRAT Nd_3 (GE) - TPM\n",
      "FRAT Ns_1 (GE) - TPM\n",
      "FRAT Ns_2 (GE) - TPM\n",
      "FRAT Ns_3 (GE) - TPM\n",
      "FRAT Oss_1 (GE) - TPM\n",
      "FRAT Oss_2 (GE) - TPM\n",
      "FRAT Oss_3 (GE) - TPM\n",
      "FRAT Oxs_1 (GE) - TPM\n",
      "FRAT Oxs_2 (GE) - TPM\n",
      "FRAT Oxs_3 (GE) - TPM\n",
      "FRAT Sp_1 (GE) - TPM\n",
      "FRAT Sp_2 (GE) - TPM\n",
      "FRAT Sp_3 (GE) - TPM\n",
      "FRAT Tm_1 (GE) - TPM\n",
      "FRAT Tm_2 (GE) - TPM\n",
      "FRAT Tm_3 (GE) - TPM\n",
      "FRAT Vic_1 (GE) - TPM\n",
      "FRAT Vic_2 (GE) - TPM\n",
      "FRAT Vic_3 (GE) - TPM\n",
      "HINF As_1 (GE) - TPM\n",
      "HINF As_2 (GE) - TPM\n",
      "HINF As_3 (GE) - TPM\n",
      "HINF Bs_1 (GE) - TPM\n",
      "HINF Bs_2 (GE) - TPM\n",
      "HINF Bs_3 (GE) - TPM\n",
      "HINF Ctrl_1 (GE) - TPM\n",
      "HINF Ctrl_2 (GE) - TPM\n",
      "HINF Ctrl_3 (GE) - TPM\n",
      "HINF Li_1 (GE) - TPM\n",
      "HINF Li_2 (GE) - TPM\n",
      "HINF Li_3 (GE) - TPM\n",
      "HINF Mig_1 (GE) - TPM\n",
      "HINF Mig_2 (GE) - TPM\n",
      "HINF Mig_3 (GE) - TPM\n",
      "HINF Nd_1 (GE) - TPM\n",
      "HINF Nd_2 (GE) - TPM\n",
      "HINF Nd_3 (GE) - TPM\n",
      "HINF Ns_1 (GE) - TPM\n",
      "HINF Ns_2 (GE) - TPM\n",
      "HINF Ns_3 (GE) - TPM\n",
      "HINF Oss_1 (GE) - TPM\n",
      "HINF Oss_2 (GE) - TPM\n",
      "HINF Oss_3 (GE) - TPM\n",
      "HINF Oxs_1 (GE) - TPM\n",
      "HINF Oxs_2 (GE) - TPM\n",
      "HINF Oxs_3 (GE) - TPM\n",
      "HINF Sp_1 (GE) - TPM\n",
      "HINF Sp_2 (GE) - TPM\n",
      "HINF Sp_3 (GE) - TPM\n",
      "HINF Tm_1 (GE) - TPM\n",
      "HINF Tm_2 (GE) - TPM\n",
      "HINF Tm_3 (GE) - TPM\n",
      "HINF Vic_1 (GE) - TPM\n",
      "HINF Vic_2 (GE) - TPM\n",
      "HINF Vic_3 (GE) - TPM\n",
      "HP G27_As_1 (GE) - TPM\n",
      "HP G27_As_2 (GE) - TPM\n",
      "HP G27_As_3 (GE) - TPM\n",
      "HP G27_Bs_1 (GE) - TPM\n",
      "HP G27_Bs_2 (GE) - TPM\n",
      "HP G27_Bs_3 (GE) - TPM\n",
      "HP G27_Ctrl_1 (GE) - TPM\n",
      "HP G27_Ctrl_2 (GE) - TPM\n",
      "HP G27_Ctrl_3 (GE) - TPM\n",
      "HP G27_Li_1 (GE) - TPM\n",
      "HP G27_Li_2 (GE) - TPM\n",
      "HP G27_Li_3 (GE) - TPM\n",
      "HP G27_Mig_1 (GE) - TPM\n",
      "HP G27_Mig_2 (GE) - TPM\n",
      "HP G27_Mig_3 (GE) - TPM\n",
      "HP G27_Nd_1 (GE) - TPM\n",
      "HP G27_Nd_2 (GE) - TPM\n",
      "HP G27_Nd_3 (GE) - TPM\n",
      "HP G27_Ns_1 (GE) - TPM\n",
      "HP G27_Ns_2 (GE) - TPM\n",
      "HP G27_Ns_3 (GE) - TPM\n",
      "HP G27_Oss_1 (GE) - TPM\n",
      "HP G27_Oss_2 (GE) - TPM\n",
      "HP G27_Oss_3 (GE) - TPM\n",
      "HP G27_Oxs_1 (GE) - TPM\n",
      "HP G27_Oxs_2 (GE) - TPM\n",
      "HP G27_Oxs_3 (GE) - TPM\n",
      "HP G27_Sp_1 (GE) - TPM\n",
      "HP G27_Sp_2 (GE) - TPM\n",
      "HP G27_Sp_3 (GE) - TPM\n",
      "HP G27_Tm_1 (GE) - TPM\n",
      "HP G27_Tm_2 (GE) - TPM\n",
      "HP G27_Tm_3 (GE) - TPM\n",
      "HP G27_Vic_1 (GE) - TPM\n",
      "HP G27_Vic_2 (GE) - TPM\n",
      "HP G27_Vic_3 (GE) - TPM\n",
      "Column names do not match across the files. Unmatched columns: {'G27_Vic_1 (GE) - TPM', 'As_2 (GE) - TPM', 'Nd_2 (GE) - TPM', 'Ctrl_1 (GE) - TPM', 'G27_Ctrl_2 (GE) - TPM', 'Oxs_1 (GE) - TPM', 'Sp_1 (GE) - TPM', 'G27_Vic_3 (GE) - TPM', 'G27_Oxs_1 (GE) - TPM', 'G27_Sp_2 (GE) - TPM', 'G27_Oss_1 (GE) - TPM', 'Ns_3 (GE) - TPM', 'Nd_3 (GE) - TPM', 'G27_Ns_2 (GE) - TPM', 'G27_Tm_2 (GE) - TPM', 'G27_As_2 (GE) - TPM', 'Tm_2 (GE) - TPM', 'Ctrl_3 (GE) - TPM', 'G27_Nd_3 (GE) - TPM', 'Tm_1 (GE) - TPM', 'G27_As_1 (GE) - TPM', 'G27_Nd_1 (GE) - TPM', 'G27_Tm_1 (GE) - TPM', 'G27_Oxs_3 (GE) - TPM', 'G27_Bs_3 (GE) - TPM', 'Ctrl_2 (GE) - TPM', 'G27_Oss_2 (GE) - TPM', 'G27_Ns_1 (GE) - TPM', 'Li_2 (GE) - TPM', 'Oss_3 (GE) - TPM', 'Bs_3 (GE) - TPM', 'Ns_1 (GE) - TPM', 'Vic_3 (GE) - TPM', 'Sp_3 (GE) - TPM', 'G27_Li_1 (GE) - TPM', 'Oss_2 (GE) - TPM', 'Bs_2 (GE) - TPM', 'G27_Bs_1 (GE) - TPM', 'G27_Ctrl_1 (GE) - TPM', 'G27_Mig_3 (GE) - TPM', 'G27_Tm_3 (GE) - TPM', 'Mig_1 (GE) - TPM', 'G27_Sp_1 (GE) - TPM', 'Nd_1 (GE) - TPM', 'Bs_1 (GE) - TPM', 'Vic_2 (GE) - TPM', 'G27_Ns_3 (GE) - TPM', 'Mig_3 (GE) - TPM', 'G27_Bs_2 (GE) - TPM', 'G27_Li_3 (GE) - TPM', 'G27_Oss_3 (GE) - TPM', 'G27_Nd_2 (GE) - TPM', 'G27_Sp_3 (GE) - TPM', 'Mig_2 (GE) - TPM', 'Li_3 (GE) - TPM', 'As_1 (GE) - TPM', 'Oss_1 (GE) - TPM', 'G27_Oxs_2 (GE) - TPM', 'Vic_1 (GE) - TPM', 'Ns_2 (GE) - TPM', 'G27_As_3 (GE) - TPM', 'G27_Li_2 (GE) - TPM', 'G27_Ctrl_3 (GE) - TPM', 'Sp_2 (GE) - TPM', 'G27_Mig_1 (GE) - TPM', 'As_3 (GE) - TPM', 'Tm_3 (GE) - TPM', 'Oxs_3 (GE) - TPM', 'Oxs_2 (GE) - TPM', 'Li_1 (GE) - TPM', 'G27_Mig_2 (GE) - TPM', 'G27_Vic_2 (GE) - TPM'}\n",
      "                                 Species   Chromosome  \\\n",
      "0      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "1      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "2      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "3      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "4      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "...                                  ...          ...   \n",
      "42886    Haemophilus influenzae 86-028NP    NC_007146   \n",
      "42887    Haemophilus influenzae 86-028NP    NC_007146   \n",
      "42888    Haemophilus influenzae 86-028NP    NC_007146   \n",
      "42889    Haemophilus influenzae 86-028NP    NC_007146   \n",
      "42890    Haemophilus influenzae 86-028NP    NC_007146   \n",
      "\n",
      "                             Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
      "0                complement(1..400)      4579.764521      2874.021275   \n",
      "1             complement(980..2143)        14.339311        17.782257   \n",
      "2            complement(2164..3555)         8.624839        10.448941   \n",
      "3            complement(3599..4984)        21.127260        38.747719   \n",
      "4            complement(5049..5855)       338.906377       344.525203   \n",
      "...                             ...              ...              ...   \n",
      "42886  complement(1908182..1908991)        77.823432        85.541198   \n",
      "42887  complement(1909021..1911102)       133.535690       120.590033   \n",
      "42888  complement(1911099..1913213)       582.971399       524.745467   \n",
      "42889  complement(1913286..1913552)      1213.062793      1076.190816   \n",
      "42890  complement(1913615..1914241)       279.604146       269.508944   \n",
      "\n",
      "       As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
      "0          2805.289041      3462.302060      3501.986142      3502.851956   \n",
      "1            23.076558        14.063192        17.260141        12.550914   \n",
      "2             3.816945         6.082623         6.523412         3.794404   \n",
      "3            23.426751        15.679651        17.443773        17.756848   \n",
      "4           239.579899       222.429542       237.845060       159.168702   \n",
      "...                ...              ...              ...              ...   \n",
      "42886       104.197363              NaN              NaN              NaN   \n",
      "42887       127.207338              NaN              NaN              NaN   \n",
      "42888       520.454167              NaN              NaN              NaN   \n",
      "42889       813.352938              NaN              NaN              NaN   \n",
      "42890       185.524463              NaN              NaN              NaN   \n",
      "\n",
      "       Ctrl_1 (GE) - TPM  ...  Sp_1 (GE) - TPM  Sp_2 (GE) - TPM  \\\n",
      "0            3906.687837  ...      7812.667199      8965.167866   \n",
      "1              22.462166  ...         4.506260        12.322023   \n",
      "2               9.456728  ...         4.306475         1.545564   \n",
      "3              43.099754  ...         4.325118         3.621928   \n",
      "4             260.879136  ...        20.427741         8.886514   \n",
      "...                  ...  ...              ...              ...   \n",
      "42886          66.192702  ...        58.155559        55.282757   \n",
      "42887          93.488720  ...        69.584948        73.544916   \n",
      "42888         504.513347  ...       287.992816       317.469375   \n",
      "42889         972.536637  ...       736.089588       816.889717   \n",
      "42890          89.833194  ...       124.383071       118.819186   \n",
      "\n",
      "       Sp_3 (GE) - TPM  Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  \\\n",
      "0          9056.547435      3205.688636      3228.682954      3103.641082   \n",
      "1             9.714290        30.958629        32.159798        33.681898   \n",
      "2             3.046184        10.307187        11.364142        15.717096   \n",
      "3             3.824213        46.222022        37.725002        48.744498   \n",
      "4            10.508767       115.769842        95.915567        75.475714   \n",
      "...                ...              ...              ...              ...   \n",
      "42886        61.742967        74.974941        73.836688        67.893015   \n",
      "42887        87.866433        73.479472        77.631662        69.297143   \n",
      "42888       377.173130       276.297803       240.806016       230.710492   \n",
      "42889       666.060019       754.936415       651.633280       630.503091   \n",
      "42890       184.978183       100.775414        83.562260        91.217027   \n",
      "\n",
      "       Vic_1 (GE) - TPM  Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \n",
      "0           3401.841238       3832.566805       3408.199122        ACHX  \n",
      "1             22.375684         22.529990         19.090084        ACHX  \n",
      "2              6.735852          5.521991          4.961554        ACHX  \n",
      "3             29.690886         27.566362         33.364651        ACHX  \n",
      "4            208.814208        333.092076        332.654255        ACHX  \n",
      "...                 ...               ...               ...         ...  \n",
      "42886        121.428625        120.086937        115.922785        HINF  \n",
      "42887        155.565124        135.125768        178.712603        HINF  \n",
      "42888        714.932779        405.341056        672.427174        HINF  \n",
      "42889       1539.291555       1462.769822       1592.252070        HINF  \n",
      "42890        352.036518        171.849807        212.992973        HINF  \n",
      "\n",
      "[42891 rows x 40 columns]\n"
     ]
    }
   ],
   "source": [
    "# Reads expression data and creates a dataframe. \n",
    "directory = 'data_expression'\n",
    "final_df = read_xlsx_files_to_dataframe(directory)\n",
    "print(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "bec2d7a5-375b-45e6-b703-874aa9754891",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[1;32mIn[69], line 2\u001B[0m\n\u001B[0;32m      1\u001B[0m directory \u001B[38;5;241m=\u001B[39m \u001B[38;5;124m'\u001B[39m\u001B[38;5;124mdata_expression\u001B[39m\u001B[38;5;124m'\u001B[39m\n\u001B[1;32m----> 2\u001B[0m final_df \u001B[38;5;241m=\u001B[39m \u001B[43mread_xlsx_files_to_dataframe\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdirectory\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mfinal_df\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m      3\u001B[0m \u001B[38;5;28mprint\u001B[39m(final_df)\n",
      "Cell \u001B[1;32mIn[46], line 29\u001B[0m, in \u001B[0;36mread_xlsx_files_to_dataframe\u001B[1;34m(directory, concatenated_df)\u001B[0m\n\u001B[0;32m     26\u001B[0m base_filename \u001B[38;5;241m=\u001B[39m os\u001B[38;5;241m.\u001B[39mpath\u001B[38;5;241m.\u001B[39msplitext(filename)[\u001B[38;5;241m0\u001B[39m]\n\u001B[0;32m     28\u001B[0m \u001B[38;5;66;03m# Read the Excel file into a pandas dataframe\u001B[39;00m\n\u001B[1;32m---> 29\u001B[0m df \u001B[38;5;241m=\u001B[39m \u001B[43mpd\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread_excel\u001B[49m\u001B[43m(\u001B[49m\u001B[43mfile_path\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m     31\u001B[0m \u001B[38;5;66;03m# Check if the species in the file is already read\u001B[39;00m\n\u001B[0;32m     32\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;124m'\u001B[39m\u001B[38;5;124mSpecies\u001B[39m\u001B[38;5;124m'\u001B[39m \u001B[38;5;129;01min\u001B[39;00m df\u001B[38;5;241m.\u001B[39mcolumns:\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\excel\\_base.py:478\u001B[0m, in \u001B[0;36mread_excel\u001B[1;34m(io, sheet_name, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, parse_dates, date_parser, date_format, thousands, decimal, comment, skipfooter, storage_options, dtype_backend)\u001B[0m\n\u001B[0;32m    476\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(io, ExcelFile):\n\u001B[0;32m    477\u001B[0m     should_close \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mTrue\u001B[39;00m\n\u001B[1;32m--> 478\u001B[0m     io \u001B[38;5;241m=\u001B[39m \u001B[43mExcelFile\u001B[49m\u001B[43m(\u001B[49m\u001B[43mio\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstorage_options\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstorage_options\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mengine\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mengine\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    479\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m engine \u001B[38;5;129;01mand\u001B[39;00m engine \u001B[38;5;241m!=\u001B[39m io\u001B[38;5;241m.\u001B[39mengine:\n\u001B[0;32m    480\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\n\u001B[0;32m    481\u001B[0m         \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mEngine should not be specified when passing \u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[0;32m    482\u001B[0m         \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124man ExcelFile - ExcelFile already has the engine set\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[0;32m    483\u001B[0m     )\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\excel\\_base.py:1513\u001B[0m, in \u001B[0;36mExcelFile.__init__\u001B[1;34m(self, path_or_buffer, engine, storage_options)\u001B[0m\n\u001B[0;32m   1510\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mengine \u001B[38;5;241m=\u001B[39m engine\n\u001B[0;32m   1511\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mstorage_options \u001B[38;5;241m=\u001B[39m storage_options\n\u001B[1;32m-> 1513\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_reader \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_engines\u001B[49m\u001B[43m[\u001B[49m\u001B[43mengine\u001B[49m\u001B[43m]\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_io\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstorage_options\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstorage_options\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\excel\\_openpyxl.py:549\u001B[0m, in \u001B[0;36mOpenpyxlReader.__init__\u001B[1;34m(self, filepath_or_buffer, storage_options)\u001B[0m\n\u001B[0;32m    539\u001B[0m \u001B[38;5;250m\u001B[39m\u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[0;32m    540\u001B[0m \u001B[38;5;124;03mReader using openpyxl engine.\u001B[39;00m\n\u001B[0;32m    541\u001B[0m \n\u001B[1;32m   (...)\u001B[0m\n\u001B[0;32m    546\u001B[0m \u001B[38;5;124;03m{storage_options}\u001B[39;00m\n\u001B[0;32m    547\u001B[0m \u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[0;32m    548\u001B[0m import_optional_dependency(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mopenpyxl\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n\u001B[1;32m--> 549\u001B[0m \u001B[38;5;28;43msuper\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[38;5;21;43m__init__\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43mfilepath_or_buffer\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstorage_options\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstorage_options\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\excel\\_base.py:540\u001B[0m, in \u001B[0;36mBaseExcelReader.__init__\u001B[1;34m(self, filepath_or_buffer, storage_options)\u001B[0m\n\u001B[0;32m    538\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mhandles\u001B[38;5;241m.\u001B[39mhandle\u001B[38;5;241m.\u001B[39mseek(\u001B[38;5;241m0\u001B[39m)\n\u001B[0;32m    539\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[1;32m--> 540\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mbook \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mload_workbook\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mhandles\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mhandle\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    541\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mException\u001B[39;00m:\n\u001B[0;32m    542\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mclose()\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\excel\\_openpyxl.py:560\u001B[0m, in \u001B[0;36mOpenpyxlReader.load_workbook\u001B[1;34m(self, filepath_or_buffer)\u001B[0m\n\u001B[0;32m    557\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mload_workbook\u001B[39m(\u001B[38;5;28mself\u001B[39m, filepath_or_buffer: FilePath \u001B[38;5;241m|\u001B[39m ReadBuffer[\u001B[38;5;28mbytes\u001B[39m]):\n\u001B[0;32m    558\u001B[0m     \u001B[38;5;28;01mfrom\u001B[39;00m \u001B[38;5;21;01mopenpyxl\u001B[39;00m \u001B[38;5;28;01mimport\u001B[39;00m load_workbook\n\u001B[1;32m--> 560\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mload_workbook\u001B[49m\u001B[43m(\u001B[49m\n\u001B[0;32m    561\u001B[0m \u001B[43m        \u001B[49m\u001B[43mfilepath_or_buffer\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mread_only\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43;01mTrue\u001B[39;49;00m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata_only\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43;01mTrue\u001B[39;49;00m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mkeep_links\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43;01mFalse\u001B[39;49;00m\n\u001B[0;32m    562\u001B[0m \u001B[43m    \u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\openpyxl\\reader\\excel.py:346\u001B[0m, in \u001B[0;36mload_workbook\u001B[1;34m(filename, read_only, keep_vba, data_only, keep_links, rich_text)\u001B[0m\n\u001B[0;32m    316\u001B[0m \u001B[38;5;250m\u001B[39m\u001B[38;5;124;03m\"\"\"Open the given filename and return the workbook\u001B[39;00m\n\u001B[0;32m    317\u001B[0m \n\u001B[0;32m    318\u001B[0m \u001B[38;5;124;03m:param filename: the path to open or a file-like object\u001B[39;00m\n\u001B[1;32m   (...)\u001B[0m\n\u001B[0;32m    342\u001B[0m \n\u001B[0;32m    343\u001B[0m \u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[0;32m    344\u001B[0m reader \u001B[38;5;241m=\u001B[39m ExcelReader(filename, read_only, keep_vba,\n\u001B[0;32m    345\u001B[0m                      data_only, keep_links, rich_text)\n\u001B[1;32m--> 346\u001B[0m \u001B[43mreader\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    347\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m reader\u001B[38;5;241m.\u001B[39mwb\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\openpyxl\\reader\\excel.py:301\u001B[0m, in \u001B[0;36mExcelReader.read\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m    299\u001B[0m apply_stylesheet(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39marchive, \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mwb)\n\u001B[0;32m    300\u001B[0m action \u001B[38;5;241m=\u001B[39m \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mread worksheets\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m--> 301\u001B[0m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread_worksheets\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    302\u001B[0m action \u001B[38;5;241m=\u001B[39m \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124massign names\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[0;32m    303\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mparser\u001B[38;5;241m.\u001B[39massign_names()\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\openpyxl\\reader\\excel.py:228\u001B[0m, in \u001B[0;36mExcelReader.read_worksheets\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m    225\u001B[0m     rels \u001B[38;5;241m=\u001B[39m get_dependents(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39marchive, rels_path)\n\u001B[0;32m    227\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mread_only:\n\u001B[1;32m--> 228\u001B[0m     ws \u001B[38;5;241m=\u001B[39m \u001B[43mReadOnlyWorksheet\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mwb\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43msheet\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mname\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrel\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mtarget\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mshared_strings\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    229\u001B[0m     ws\u001B[38;5;241m.\u001B[39msheet_state \u001B[38;5;241m=\u001B[39m sheet\u001B[38;5;241m.\u001B[39mstate\n\u001B[0;32m    230\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mwb\u001B[38;5;241m.\u001B[39m_sheets\u001B[38;5;241m.\u001B[39mappend(ws)\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\openpyxl\\worksheet\\_read_only.py:42\u001B[0m, in \u001B[0;36mReadOnlyWorksheet.__init__\u001B[1;34m(self, parent_workbook, title, worksheet_path, shared_strings)\u001B[0m\n\u001B[0;32m     40\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_worksheet_path \u001B[38;5;241m=\u001B[39m worksheet_path\n\u001B[0;32m     41\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_shared_strings \u001B[38;5;241m=\u001B[39m shared_strings\n\u001B[1;32m---> 42\u001B[0m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_get_size\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m     43\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mdefined_names \u001B[38;5;241m=\u001B[39m DefinedNameDict()\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\openpyxl\\worksheet\\_read_only.py:49\u001B[0m, in \u001B[0;36mReadOnlyWorksheet._get_size\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m     47\u001B[0m src \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_get_source()\n\u001B[0;32m     48\u001B[0m parser \u001B[38;5;241m=\u001B[39m WorkSheetParser(src, [])\n\u001B[1;32m---> 49\u001B[0m dimensions \u001B[38;5;241m=\u001B[39m \u001B[43mparser\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mparse_dimensions\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m     50\u001B[0m src\u001B[38;5;241m.\u001B[39mclose()\n\u001B[0;32m     51\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m dimensions \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\openpyxl\\worksheet\\_reader.py:178\u001B[0m, in \u001B[0;36mWorkSheetParser.parse_dimensions\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m    173\u001B[0m \u001B[38;5;250m\u001B[39m\u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[0;32m    174\u001B[0m \u001B[38;5;124;03mGet worksheet dimensions if they are provided.\u001B[39;00m\n\u001B[0;32m    175\u001B[0m \u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[0;32m    176\u001B[0m it \u001B[38;5;241m=\u001B[39m iterparse(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39msource)\n\u001B[1;32m--> 178\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m _event, element \u001B[38;5;129;01min\u001B[39;00m it:\n\u001B[0;32m    179\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m element\u001B[38;5;241m.\u001B[39mtag \u001B[38;5;241m==\u001B[39m DIMENSION_TAG:\n\u001B[0;32m    180\u001B[0m         dim \u001B[38;5;241m=\u001B[39m SheetDimension\u001B[38;5;241m.\u001B[39mfrom_tree(element)\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\xml\\etree\\ElementTree.py:1251\u001B[0m, in \u001B[0;36miterparse.<locals>.iterator\u001B[1;34m(source)\u001B[0m\n\u001B[0;32m   1249\u001B[0m \u001B[38;5;28;01myield from\u001B[39;00m pullparser\u001B[38;5;241m.\u001B[39mread_events()\n\u001B[0;32m   1250\u001B[0m \u001B[38;5;66;03m# load event buffer\u001B[39;00m\n\u001B[1;32m-> 1251\u001B[0m data \u001B[38;5;241m=\u001B[39m \u001B[43msource\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;241;43m16\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;241;43m1024\u001B[39;49m\u001B[43m)\u001B[49m\n\u001B[0;32m   1252\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m data:\n\u001B[0;32m   1253\u001B[0m     \u001B[38;5;28;01mbreak\u001B[39;00m\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\zipfile.py:951\u001B[0m, in \u001B[0;36mZipExtFile.read\u001B[1;34m(self, n)\u001B[0m\n\u001B[0;32m    949\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_offset \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m0\u001B[39m\n\u001B[0;32m    950\u001B[0m \u001B[38;5;28;01mwhile\u001B[39;00m n \u001B[38;5;241m>\u001B[39m \u001B[38;5;241m0\u001B[39m \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_eof:\n\u001B[1;32m--> 951\u001B[0m     data \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_read1\u001B[49m\u001B[43m(\u001B[49m\u001B[43mn\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    952\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m n \u001B[38;5;241m<\u001B[39m \u001B[38;5;28mlen\u001B[39m(data):\n\u001B[0;32m    953\u001B[0m         \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_readbuffer \u001B[38;5;241m=\u001B[39m data\n",
      "File \u001B[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\zipfile.py:1027\u001B[0m, in \u001B[0;36mZipExtFile._read1\u001B[1;34m(self, n)\u001B[0m\n\u001B[0;32m   1025\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_compress_type \u001B[38;5;241m==\u001B[39m ZIP_DEFLATED:\n\u001B[0;32m   1026\u001B[0m     n \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mmax\u001B[39m(n, \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mMIN_READ_SIZE)\n\u001B[1;32m-> 1027\u001B[0m     data \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_decompressor\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mdecompress\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdata\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mn\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m   1028\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_eof \u001B[38;5;241m=\u001B[39m (\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_decompressor\u001B[38;5;241m.\u001B[39meof \u001B[38;5;129;01mor\u001B[39;00m\n\u001B[0;32m   1029\u001B[0m                  \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_compress_left \u001B[38;5;241m<\u001B[39m\u001B[38;5;241m=\u001B[39m \u001B[38;5;241m0\u001B[39m \u001B[38;5;129;01mand\u001B[39;00m\n\u001B[0;32m   1030\u001B[0m                  \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_decompressor\u001B[38;5;241m.\u001B[39munconsumed_tail)\n\u001B[0;32m   1031\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_eof:\n",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "source": [
    "# Run this cell if read_xlsx_files_to_dataframe is interrupted or raised an error. In the next run error is handled. \n",
    "# You can subset the species in another directory for the next run to increase the speed up.\n",
    "\n",
    "# Uncomment below to continue to append to final_df\n",
    "\n",
    "# directory = 'data_expression'\n",
    "# final_df = read_xlsx_files_to_dataframe(directory, final_df)\n",
    "# print(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1ce933e0-aa0e-4ff2-b809-d47fe5f0a4ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>As_1 (GE) - TPM</th>\n",
       "      <th>As_2 (GE) - TPM</th>\n",
       "      <th>As_3 (GE) - TPM</th>\n",
       "      <th>Bs_1 (GE) - TPM</th>\n",
       "      <th>Bs_2 (GE) - TPM</th>\n",
       "      <th>Bs_3 (GE) - TPM</th>\n",
       "      <th>Ctrl_1 (GE) - TPM</th>\n",
       "      <th>...</th>\n",
       "      <th>Sp_1 (GE) - TPM</th>\n",
       "      <th>Sp_2 (GE) - TPM</th>\n",
       "      <th>Sp_3 (GE) - TPM</th>\n",
       "      <th>Tm_1 (GE) - TPM</th>\n",
       "      <th>Tm_2 (GE) - TPM</th>\n",
       "      <th>Tm_3 (GE) - TPM</th>\n",
       "      <th>Vic_1 (GE) - TPM</th>\n",
       "      <th>Vic_2 (GE) - TPM</th>\n",
       "      <th>Vic_3 (GE) - TPM</th>\n",
       "      <th>Species ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(1..400)</td>\n",
       "      <td>4579.764521</td>\n",
       "      <td>2874.021275</td>\n",
       "      <td>2805.289041</td>\n",
       "      <td>3462.302060</td>\n",
       "      <td>3501.986142</td>\n",
       "      <td>3502.851956</td>\n",
       "      <td>3906.687837</td>\n",
       "      <td>...</td>\n",
       "      <td>7812.667199</td>\n",
       "      <td>8965.167866</td>\n",
       "      <td>9056.547435</td>\n",
       "      <td>3205.688636</td>\n",
       "      <td>3228.682954</td>\n",
       "      <td>3103.641082</td>\n",
       "      <td>3401.841238</td>\n",
       "      <td>3832.566805</td>\n",
       "      <td>3408.199122</td>\n",
       "      <td>ACHX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(980..2143)</td>\n",
       "      <td>14.339311</td>\n",
       "      <td>17.782257</td>\n",
       "      <td>23.076558</td>\n",
       "      <td>14.063192</td>\n",
       "      <td>17.260141</td>\n",
       "      <td>12.550914</td>\n",
       "      <td>22.462166</td>\n",
       "      <td>...</td>\n",
       "      <td>4.506260</td>\n",
       "      <td>12.322023</td>\n",
       "      <td>9.714290</td>\n",
       "      <td>30.958629</td>\n",
       "      <td>32.159798</td>\n",
       "      <td>33.681898</td>\n",
       "      <td>22.375684</td>\n",
       "      <td>22.529990</td>\n",
       "      <td>19.090084</td>\n",
       "      <td>ACHX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(2164..3555)</td>\n",
       "      <td>8.624839</td>\n",
       "      <td>10.448941</td>\n",
       "      <td>3.816945</td>\n",
       "      <td>6.082623</td>\n",
       "      <td>6.523412</td>\n",
       "      <td>3.794404</td>\n",
       "      <td>9.456728</td>\n",
       "      <td>...</td>\n",
       "      <td>4.306475</td>\n",
       "      <td>1.545564</td>\n",
       "      <td>3.046184</td>\n",
       "      <td>10.307187</td>\n",
       "      <td>11.364142</td>\n",
       "      <td>15.717096</td>\n",
       "      <td>6.735852</td>\n",
       "      <td>5.521991</td>\n",
       "      <td>4.961554</td>\n",
       "      <td>ACHX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(3599..4984)</td>\n",
       "      <td>21.127260</td>\n",
       "      <td>38.747719</td>\n",
       "      <td>23.426751</td>\n",
       "      <td>15.679651</td>\n",
       "      <td>17.443773</td>\n",
       "      <td>17.756848</td>\n",
       "      <td>43.099754</td>\n",
       "      <td>...</td>\n",
       "      <td>4.325118</td>\n",
       "      <td>3.621928</td>\n",
       "      <td>3.824213</td>\n",
       "      <td>46.222022</td>\n",
       "      <td>37.725002</td>\n",
       "      <td>48.744498</td>\n",
       "      <td>29.690886</td>\n",
       "      <td>27.566362</td>\n",
       "      <td>33.364651</td>\n",
       "      <td>ACHX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(5049..5855)</td>\n",
       "      <td>338.906377</td>\n",
       "      <td>344.525203</td>\n",
       "      <td>239.579899</td>\n",
       "      <td>222.429542</td>\n",
       "      <td>237.845060</td>\n",
       "      <td>159.168702</td>\n",
       "      <td>260.879136</td>\n",
       "      <td>...</td>\n",
       "      <td>20.427741</td>\n",
       "      <td>8.886514</td>\n",
       "      <td>10.508767</td>\n",
       "      <td>115.769842</td>\n",
       "      <td>95.915567</td>\n",
       "      <td>75.475714</td>\n",
       "      <td>208.814208</td>\n",
       "      <td>333.092076</td>\n",
       "      <td>332.654255</td>\n",
       "      <td>ACHX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105083</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>2000927..2003506</td>\n",
       "      <td>154.651786</td>\n",
       "      <td>166.363005</td>\n",
       "      <td>158.597735</td>\n",
       "      <td>103.096554</td>\n",
       "      <td>117.077900</td>\n",
       "      <td>113.354978</td>\n",
       "      <td>153.575746</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>186.116702</td>\n",
       "      <td>195.310415</td>\n",
       "      <td>191.215538</td>\n",
       "      <td>121.440741</td>\n",
       "      <td>149.782243</td>\n",
       "      <td>68.981909</td>\n",
       "      <td>SSUIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105084</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>complement(2003907..2004590)</td>\n",
       "      <td>51.843037</td>\n",
       "      <td>47.737511</td>\n",
       "      <td>50.150766</td>\n",
       "      <td>30.035985</td>\n",
       "      <td>27.269205</td>\n",
       "      <td>22.901773</td>\n",
       "      <td>142.767970</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69.873527</td>\n",
       "      <td>55.227042</td>\n",
       "      <td>67.758348</td>\n",
       "      <td>94.401511</td>\n",
       "      <td>108.957265</td>\n",
       "      <td>44.032987</td>\n",
       "      <td>SSUIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105085</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>complement(2004615..2005094)</td>\n",
       "      <td>72.526107</td>\n",
       "      <td>104.306461</td>\n",
       "      <td>90.287062</td>\n",
       "      <td>67.168967</td>\n",
       "      <td>49.922529</td>\n",
       "      <td>76.483851</td>\n",
       "      <td>292.305112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>161.530080</td>\n",
       "      <td>151.730774</td>\n",
       "      <td>194.563257</td>\n",
       "      <td>245.576984</td>\n",
       "      <td>273.601934</td>\n",
       "      <td>308.030757</td>\n",
       "      <td>SSUIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105086</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>2005268..2006464</td>\n",
       "      <td>133.774758</td>\n",
       "      <td>140.802925</td>\n",
       "      <td>145.617949</td>\n",
       "      <td>191.050633</td>\n",
       "      <td>234.493519</td>\n",
       "      <td>234.235125</td>\n",
       "      <td>136.907216</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180.352408</td>\n",
       "      <td>159.162080</td>\n",
       "      <td>162.029477</td>\n",
       "      <td>143.215049</td>\n",
       "      <td>132.640597</td>\n",
       "      <td>134.195769</td>\n",
       "      <td>SSUIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105087</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>2006525..2007289</td>\n",
       "      <td>322.297911</td>\n",
       "      <td>297.998136</td>\n",
       "      <td>289.298762</td>\n",
       "      <td>276.320634</td>\n",
       "      <td>306.466652</td>\n",
       "      <td>321.135201</td>\n",
       "      <td>448.980652</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>300.848584</td>\n",
       "      <td>373.308813</td>\n",
       "      <td>348.861950</td>\n",
       "      <td>313.318233</td>\n",
       "      <td>372.073690</td>\n",
       "      <td>286.332148</td>\n",
       "      <td>SSUIS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105088 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Species   Chromosome  \\\n",
       "0       Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "1       Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "2       Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "3       Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "4       Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "...                                   ...          ...   \n",
       "105083       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "105084       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "105085       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "105086       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "105087       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "\n",
       "                              Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
       "0                 complement(1..400)      4579.764521      2874.021275   \n",
       "1              complement(980..2143)        14.339311        17.782257   \n",
       "2             complement(2164..3555)         8.624839        10.448941   \n",
       "3             complement(3599..4984)        21.127260        38.747719   \n",
       "4             complement(5049..5855)       338.906377       344.525203   \n",
       "...                              ...              ...              ...   \n",
       "105083              2000927..2003506       154.651786       166.363005   \n",
       "105084  complement(2003907..2004590)        51.843037        47.737511   \n",
       "105085  complement(2004615..2005094)        72.526107       104.306461   \n",
       "105086              2005268..2006464       133.774758       140.802925   \n",
       "105087              2006525..2007289       322.297911       297.998136   \n",
       "\n",
       "        As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
       "0           2805.289041      3462.302060      3501.986142      3502.851956   \n",
       "1             23.076558        14.063192        17.260141        12.550914   \n",
       "2              3.816945         6.082623         6.523412         3.794404   \n",
       "3             23.426751        15.679651        17.443773        17.756848   \n",
       "4            239.579899       222.429542       237.845060       159.168702   \n",
       "...                 ...              ...              ...              ...   \n",
       "105083       158.597735       103.096554       117.077900       113.354978   \n",
       "105084        50.150766        30.035985        27.269205        22.901773   \n",
       "105085        90.287062        67.168967        49.922529        76.483851   \n",
       "105086       145.617949       191.050633       234.493519       234.235125   \n",
       "105087       289.298762       276.320634       306.466652       321.135201   \n",
       "\n",
       "        Ctrl_1 (GE) - TPM  ...  Sp_1 (GE) - TPM  Sp_2 (GE) - TPM  \\\n",
       "0             3906.687837  ...      7812.667199      8965.167866   \n",
       "1               22.462166  ...         4.506260        12.322023   \n",
       "2                9.456728  ...         4.306475         1.545564   \n",
       "3               43.099754  ...         4.325118         3.621928   \n",
       "4              260.879136  ...        20.427741         8.886514   \n",
       "...                   ...  ...              ...              ...   \n",
       "105083         153.575746  ...              NaN              NaN   \n",
       "105084         142.767970  ...              NaN              NaN   \n",
       "105085         292.305112  ...              NaN              NaN   \n",
       "105086         136.907216  ...              NaN              NaN   \n",
       "105087         448.980652  ...              NaN              NaN   \n",
       "\n",
       "        Sp_3 (GE) - TPM  Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  \\\n",
       "0           9056.547435      3205.688636      3228.682954      3103.641082   \n",
       "1              9.714290        30.958629        32.159798        33.681898   \n",
       "2              3.046184        10.307187        11.364142        15.717096   \n",
       "3              3.824213        46.222022        37.725002        48.744498   \n",
       "4             10.508767       115.769842        95.915567        75.475714   \n",
       "...                 ...              ...              ...              ...   \n",
       "105083              NaN       186.116702       195.310415       191.215538   \n",
       "105084              NaN        69.873527        55.227042        67.758348   \n",
       "105085              NaN       161.530080       151.730774       194.563257   \n",
       "105086              NaN       180.352408       159.162080       162.029477   \n",
       "105087              NaN       300.848584       373.308813       348.861950   \n",
       "\n",
       "        Vic_1 (GE) - TPM  Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \n",
       "0            3401.841238       3832.566805       3408.199122        ACHX  \n",
       "1              22.375684         22.529990         19.090084        ACHX  \n",
       "2               6.735852          5.521991          4.961554        ACHX  \n",
       "3              29.690886         27.566362         33.364651        ACHX  \n",
       "4             208.814208        333.092076        332.654255        ACHX  \n",
       "...                  ...               ...               ...         ...  \n",
       "105083        121.440741        149.782243         68.981909       SSUIS  \n",
       "105084         94.401511        108.957265         44.032987       SSUIS  \n",
       "105085        245.576984        273.601934        308.030757       SSUIS  \n",
       "105086        143.215049        132.640597        134.195769       SSUIS  \n",
       "105087        313.318233        372.073690        286.332148       SSUIS  \n",
       "\n",
       "[105088 rows x 40 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "98c6d709-bc23-4ab9-85cd-e13fe1588fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv('expression_final_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f0294b83-f340-4c4c-8d95-0df59f680daf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species:  32\n",
      "{'Enterococcus faecalis OG1RF', 'Mycobacterium tuberculosis H37Ra', 'Streptococcus agalactiae NEM316', 'Neisseria meningitidis serogroup C FAM18', 'Francisella tularensis subsp. holarctica FSC200', 'Streptococcus suis S10 -P1/7', 'Pseudomonas aeruginosa PAO1', 'Escherichia coli UPEC 536', 'Legionella pneumophila subsp. pneumophila Philadelphia 1', 'Streptococcus pyogenes 5448', 'Helicobacter pylori G27', 'Acinetobacter baumannii AB5075-UW', 'Shigella flexneri 5a str. M90T', 'Staphylococcus\\xa0epidermidis 1457', 'Burkholderia pseudomallei K96243', 'Listeria monocytogenes EGD-e', 'Haemophilus influenzae 86-028NP', 'Helicobacter pylori J99', 'Staphylococcus\\xa0aureus MSSA476', 'Achromobacter xylosoxidans SOLR10', 'Neisseria gonorrhoeae FA 1090', 'Streptococcus pneumoniae D39', 'Yersinia pseudotuberculosis YPIII', 'Vibrio cholerae O1 biovar El Tor str. N16961', 'Escherichia coli EPEC 0127:H6 E2348/69', 'Borrelia burgdorferi B31', 'Aggregatibacter actinomycetemcomitans D7S-1', 'Klebsiella pneumoniae subsp. pneumoniae MGH 78578', 'Escherichia coli ETEC H10407', 'Staphylococcus\\xa0aureus MRSA252', 'Salmonella enterica subsp. enterica serovar Typhimurium SL1344', 'Campylobacter jejuni subsp. jejuni 81-176'}\n"
     ]
    }
   ],
   "source": [
    "print('Number of species: ', len(set(final_df[\"Species\"].tolist())))\n",
    "print(set(final_df[\"Species\"].tolist()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3e2724a-adde-4dc1-9704-164d48d9e2ed",
   "metadata": {},
   "source": [
    "# upstream_sequences section (Move after you get all the species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e3867fe2-b4fa-46bc-bb20-059c5efceb84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions to process sequence data\n",
    "def preprocess_tsv(tsv_path):\n",
    "    # Read the TSV file into a pandas dataframe\n",
    "    tsv_df = pd.read_csv(tsv_path, sep='\\t')\n",
    "    \n",
    "    # Drop the .csv extension from the 'csv' column and rename it to 'Species'\n",
    "    tsv_df['Species'] = tsv_df['csv'].str.replace('.csv', '', regex=False)\n",
    "    \n",
    "    # Rename the 'contig' column to 'Chromosome'\n",
    "    tsv_df.rename(columns={'contig': 'Chromosome', 'region': 'Region'}, inplace=True)\n",
    "    \n",
    "    # Drop the original 'csv' column\n",
    "    tsv_df.drop(columns=['csv'], inplace=True)\n",
    "\n",
    "    # Replace specific species names\n",
    "    tsv_df['Species'] = tsv_df['Species'].replace({\n",
    "        'Salmonella enterica subsp. enterica serovar TyphimuriumSL1344': 'Salmonella enterica subsp. enterica serovar Typhimurium SL1344',\n",
    "        'Escherichia coli EPEC 0127 H6 E2348 69': 'Escherichia coli EPEC 0127:H6 E2348/69',\n",
    "        'Streptococcus suis S10 P 17': 'Streptococcus suis S10 -P1/7'\n",
    "    })\n",
    "    \n",
    "    return tsv_df\n",
    "\n",
    "def merge_dataframes(main_df, tsv_df):\n",
    "    # Check if the two datasets have the same species\n",
    "    species_diff_1 = set(tsv_df['Species']).difference(set(main_df['Species']))\n",
    "    species_diff_2 = set(main_df['Species']).difference(set(tsv_df['Species']))\n",
    "\n",
    "    if species_diff_1 or species_diff_2:\n",
    "        print(f\"Species present in TSV but not in main dataframe: {species_diff_1}\")\n",
    "        print(f\"Species present in main dataframe but not in TSV: {species_diff_2}\")\n",
    "\n",
    "    # Perform an inner merge on 'Species', 'Chromosome', and 'Region' columns\n",
    "    # What other merge can we do here!!\n",
    "    merged_df = pd.merge(main_df, tsv_df, on=['Species', 'Chromosome', 'Region'], how='inner')\n",
    "    \n",
    "    return merged_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "89b402ce-6af9-4388-961e-343abfc1161e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read and preprocess the TSV file\n",
    "tsv_path = 'upstream_sequences.tsv'\n",
    "sequence_df = preprocess_tsv(tsv_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7b20ac7a-48d4-49f5-87f1-0e763c90c895",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>upstream200</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(235..402)</td>\n",
       "      <td>CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(372..806)</td>\n",
       "      <td>ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(816..2210)</td>\n",
       "      <td>TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(2271..3896)</td>\n",
       "      <td>TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_002505</td>\n",
       "      <td>complement(3899..4156)</td>\n",
       "      <td>CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96406</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3393934..3395274)</td>\n",
       "      <td>CAGCACCAGCTGATCCAATGCAAGCTAAGGTAATGATGTTTTTACC...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96407</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3395275..3396945)</td>\n",
       "      <td>ATCAGTATTTTATTAGCCCTTTGATAACACCATGTTGTCGCTATTA...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96408</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3396955..3397200)</td>\n",
       "      <td>AATAAGTTAGGCTATGCACGCCTTGGTTTAGCATTGTCAAAAAAAA...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96409</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3397167..3397355)</td>\n",
       "      <td>TAAAAAGACGTCGTGCTAAAGGTCGTAAGCGTTTATCTGCCTAAGT...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96410</th>\n",
       "      <td>NC_002942</td>\n",
       "      <td>complement(3397512..3397646)</td>\n",
       "      <td>CGACCTGTAAATAAAAAATTACACAAAAAATAATCAGTTATTCAAA...</td>\n",
       "      <td>Legionella pneumophila subsp. pneumophila Phil...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96411 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome                        Region  \\\n",
       "0      NC_002505          complement(235..402)   \n",
       "1      NC_002505          complement(372..806)   \n",
       "2      NC_002505         complement(816..2210)   \n",
       "3      NC_002505        complement(2271..3896)   \n",
       "4      NC_002505        complement(3899..4156)   \n",
       "...          ...                           ...   \n",
       "96406  NC_002942  complement(3393934..3395274)   \n",
       "96407  NC_002942  complement(3395275..3396945)   \n",
       "96408  NC_002942  complement(3396955..3397200)   \n",
       "96409  NC_002942  complement(3397167..3397355)   \n",
       "96410  NC_002942  complement(3397512..3397646)   \n",
       "\n",
       "                                             upstream200  \\\n",
       "0      CAGGCTCTGCAGAATACACCACCGAATACCTCTGCACTACGTTATG...   \n",
       "1      ATCTCGATGCCCTAGAGCGAGCCGCAGAGCACTTAGCGATTGGCCA...   \n",
       "2      TGCCAGTCATGTTCACTTTCTTCTTCCTGTGGTTCCCATCAGGTCT...   \n",
       "3      TGGTTTATTAGTCCACTTATCGGCCCACGCTGCCGATTCACTCCTA...   \n",
       "4      CTTTCTCATCCTCGTTTGGGACTCGCGGTTCCTAAAAAGCAGATCA...   \n",
       "...                                                  ...   \n",
       "96406  CAGCACCAGCTGATCCAATGCAAGCTAAGGTAATGATGTTTTTACC...   \n",
       "96407  ATCAGTATTTTATTAGCCCTTTGATAACACCATGTTGTCGCTATTA...   \n",
       "96408  AATAAGTTAGGCTATGCACGCCTTGGTTTAGCATTGTCAAAAAAAA...   \n",
       "96409  TAAAAAGACGTCGTGCTAAAGGTCGTAAGCGTTTATCTGCCTAAGT...   \n",
       "96410  CGACCTGTAAATAAAAAATTACACAAAAAATAATCAGTTATTCAAA...   \n",
       "\n",
       "                                                 Species  \n",
       "0           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "1           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "2           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "3           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "4           Vibrio cholerae O1 biovar El Tor str. N16961  \n",
       "...                                                  ...  \n",
       "96406  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96407  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96408  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96409  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "96410  Legionella pneumophila subsp. pneumophila Phil...  \n",
       "\n",
       "[96411 rows x 4 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4d35cf33-bc52-4808-9c18-66023e79a992",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species present in TSV but not in main dataframe: set()\n",
      "Species present in main dataframe but not in TSV: {'Shigella flexneri 5a str. M90T', 'Yersinia pseudotuberculosis YPIII'}\n",
      "                                 Species   Chromosome  \\\n",
      "0      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "1      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "2      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "3      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "4      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
      "...                                  ...          ...   \n",
      "96406       Streptococcus suis S10 -P1/7    NC_012925   \n",
      "96407       Streptococcus suis S10 -P1/7    NC_012925   \n",
      "96408       Streptococcus suis S10 -P1/7    NC_012925   \n",
      "96409       Streptococcus suis S10 -P1/7    NC_012925   \n",
      "96410       Streptococcus suis S10 -P1/7    NC_012925   \n",
      "\n",
      "                             Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
      "0                complement(1..400)      4579.764521      2874.021275   \n",
      "1             complement(980..2143)        14.339311        17.782257   \n",
      "2            complement(2164..3555)         8.624839        10.448941   \n",
      "3            complement(3599..4984)        21.127260        38.747719   \n",
      "4            complement(5049..5855)       338.906377       344.525203   \n",
      "...                             ...              ...              ...   \n",
      "96406              2000927..2003506       154.651786       166.363005   \n",
      "96407  complement(2003907..2004590)        51.843037        47.737511   \n",
      "96408  complement(2004615..2005094)        72.526107       104.306461   \n",
      "96409              2005268..2006464       133.774758       140.802925   \n",
      "96410              2006525..2007289       322.297911       297.998136   \n",
      "\n",
      "       As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
      "0          2805.289041      3462.302060      3501.986142      3502.851956   \n",
      "1            23.076558        14.063192        17.260141        12.550914   \n",
      "2             3.816945         6.082623         6.523412         3.794404   \n",
      "3            23.426751        15.679651        17.443773        17.756848   \n",
      "4           239.579899       222.429542       237.845060       159.168702   \n",
      "...                ...              ...              ...              ...   \n",
      "96406       158.597735       103.096554       117.077900       113.354978   \n",
      "96407        50.150766        30.035985        27.269205        22.901773   \n",
      "96408        90.287062        67.168967        49.922529        76.483851   \n",
      "96409       145.617949       191.050633       234.493519       234.235125   \n",
      "96410       289.298762       276.320634       306.466652       321.135201   \n",
      "\n",
      "       Ctrl_1 (GE) - TPM  ...  Sp_2 (GE) - TPM  Sp_3 (GE) - TPM  \\\n",
      "0            3906.687837  ...      8965.167866      9056.547435   \n",
      "1              22.462166  ...        12.322023         9.714290   \n",
      "2               9.456728  ...         1.545564         3.046184   \n",
      "3              43.099754  ...         3.621928         3.824213   \n",
      "4             260.879136  ...         8.886514        10.508767   \n",
      "...                  ...  ...              ...              ...   \n",
      "96406         153.575746  ...              NaN              NaN   \n",
      "96407         142.767970  ...              NaN              NaN   \n",
      "96408         292.305112  ...              NaN              NaN   \n",
      "96409         136.907216  ...              NaN              NaN   \n",
      "96410         448.980652  ...              NaN              NaN   \n",
      "\n",
      "       Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  Vic_1 (GE) - TPM  \\\n",
      "0          3205.688636      3228.682954      3103.641082       3401.841238   \n",
      "1            30.958629        32.159798        33.681898         22.375684   \n",
      "2            10.307187        11.364142        15.717096          6.735852   \n",
      "3            46.222022        37.725002        48.744498         29.690886   \n",
      "4           115.769842        95.915567        75.475714        208.814208   \n",
      "...                ...              ...              ...               ...   \n",
      "96406       186.116702       195.310415       191.215538        121.440741   \n",
      "96407        69.873527        55.227042        67.758348         94.401511   \n",
      "96408       161.530080       151.730774       194.563257        245.576984   \n",
      "96409       180.352408       159.162080       162.029477        143.215049   \n",
      "96410       300.848584       373.308813       348.861950        313.318233   \n",
      "\n",
      "       Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \\\n",
      "0           3832.566805       3408.199122        ACHX   \n",
      "1             22.529990         19.090084        ACHX   \n",
      "2              5.521991          4.961554        ACHX   \n",
      "3             27.566362         33.364651        ACHX   \n",
      "4            333.092076        332.654255        ACHX   \n",
      "...                 ...               ...         ...   \n",
      "96406        149.782243         68.981909       SSUIS   \n",
      "96407        108.957265         44.032987       SSUIS   \n",
      "96408        273.601934        308.030757       SSUIS   \n",
      "96409        132.640597        134.195769       SSUIS   \n",
      "96410        372.073690        286.332148       SSUIS   \n",
      "\n",
      "                                             upstream200  \n",
      "0      CCTTCCAAGCTTACGACGAGGGTTCGATTCCCTTCACCCGCTCCAA...  \n",
      "1      GCATTGCATCTCGGCCGCCAGTTGCGCACGGGCACCGTCTTCATGA...  \n",
      "2      GGAAGCGGCGCAGCGCCTGCGCCAGTCCGAAGCCGCGCGCGAGCTG...  \n",
      "3      CGAAGGCGCGGGCGGCGTCATTACCGATTGGTCGGGGCAGCCGTTA...  \n",
      "4      AAGGAATACACGGCAACGGCGATGAGCTGCTTTTCTGCGGACGGCG...  \n",
      "...                                                  ...  \n",
      "96406  TGGCCATTACGACTGTTGCGAACAAAGTTCGCCTTATCTCCAACCT...  \n",
      "96407  ATATAACTTTTATCATCGGTGGTAGTTTGGGTCTCGATTCTTGTAT...  \n",
      "96408  TTTTTTCATATAAGGAATCCTTTCCATTTCTTTAATTTTCCTTAAG...  \n",
      "96409  TCCAACGGTAATCAATTTTATTTTCATAAAATAATTGTAACATATC...  \n",
      "96410  TTCAAAATAAGAGTGACCTCCAGAGCGCTCTCTACAAACATCAAAT...  \n",
      "\n",
      "[96411 rows x 41 columns]\n"
     ]
    }
   ],
   "source": [
    "merged_df = merge_dataframes(final_df, sequence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "8dc7950f-e119-44ad-9abd-f7ba476e0eb7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Region</th>\n",
       "      <th>As_1 (GE) - TPM</th>\n",
       "      <th>As_2 (GE) - TPM</th>\n",
       "      <th>As_3 (GE) - TPM</th>\n",
       "      <th>Bs_1 (GE) - TPM</th>\n",
       "      <th>Bs_2 (GE) - TPM</th>\n",
       "      <th>Bs_3 (GE) - TPM</th>\n",
       "      <th>Ctrl_1 (GE) - TPM</th>\n",
       "      <th>...</th>\n",
       "      <th>Sp_2 (GE) - TPM</th>\n",
       "      <th>Sp_3 (GE) - TPM</th>\n",
       "      <th>Tm_1 (GE) - TPM</th>\n",
       "      <th>Tm_2 (GE) - TPM</th>\n",
       "      <th>Tm_3 (GE) - TPM</th>\n",
       "      <th>Vic_1 (GE) - TPM</th>\n",
       "      <th>Vic_2 (GE) - TPM</th>\n",
       "      <th>Vic_3 (GE) - TPM</th>\n",
       "      <th>Species ID</th>\n",
       "      <th>upstream200</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(1..400)</td>\n",
       "      <td>4579.764521</td>\n",
       "      <td>2874.021275</td>\n",
       "      <td>2805.289041</td>\n",
       "      <td>3462.302060</td>\n",
       "      <td>3501.986142</td>\n",
       "      <td>3502.851956</td>\n",
       "      <td>3906.687837</td>\n",
       "      <td>...</td>\n",
       "      <td>8965.167866</td>\n",
       "      <td>9056.547435</td>\n",
       "      <td>3205.688636</td>\n",
       "      <td>3228.682954</td>\n",
       "      <td>3103.641082</td>\n",
       "      <td>3401.841238</td>\n",
       "      <td>3832.566805</td>\n",
       "      <td>3408.199122</td>\n",
       "      <td>ACHX</td>\n",
       "      <td>CCTTCCAAGCTTACGACGAGGGTTCGATTCCCTTCACCCGCTCCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(980..2143)</td>\n",
       "      <td>14.339311</td>\n",
       "      <td>17.782257</td>\n",
       "      <td>23.076558</td>\n",
       "      <td>14.063192</td>\n",
       "      <td>17.260141</td>\n",
       "      <td>12.550914</td>\n",
       "      <td>22.462166</td>\n",
       "      <td>...</td>\n",
       "      <td>12.322023</td>\n",
       "      <td>9.714290</td>\n",
       "      <td>30.958629</td>\n",
       "      <td>32.159798</td>\n",
       "      <td>33.681898</td>\n",
       "      <td>22.375684</td>\n",
       "      <td>22.529990</td>\n",
       "      <td>19.090084</td>\n",
       "      <td>ACHX</td>\n",
       "      <td>GCATTGCATCTCGGCCGCCAGTTGCGCACGGGCACCGTCTTCATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(2164..3555)</td>\n",
       "      <td>8.624839</td>\n",
       "      <td>10.448941</td>\n",
       "      <td>3.816945</td>\n",
       "      <td>6.082623</td>\n",
       "      <td>6.523412</td>\n",
       "      <td>3.794404</td>\n",
       "      <td>9.456728</td>\n",
       "      <td>...</td>\n",
       "      <td>1.545564</td>\n",
       "      <td>3.046184</td>\n",
       "      <td>10.307187</td>\n",
       "      <td>11.364142</td>\n",
       "      <td>15.717096</td>\n",
       "      <td>6.735852</td>\n",
       "      <td>5.521991</td>\n",
       "      <td>4.961554</td>\n",
       "      <td>ACHX</td>\n",
       "      <td>GGAAGCGGCGCAGCGCCTGCGCCAGTCCGAAGCCGCGCGCGAGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(3599..4984)</td>\n",
       "      <td>21.127260</td>\n",
       "      <td>38.747719</td>\n",
       "      <td>23.426751</td>\n",
       "      <td>15.679651</td>\n",
       "      <td>17.443773</td>\n",
       "      <td>17.756848</td>\n",
       "      <td>43.099754</td>\n",
       "      <td>...</td>\n",
       "      <td>3.621928</td>\n",
       "      <td>3.824213</td>\n",
       "      <td>46.222022</td>\n",
       "      <td>37.725002</td>\n",
       "      <td>48.744498</td>\n",
       "      <td>29.690886</td>\n",
       "      <td>27.566362</td>\n",
       "      <td>33.364651</td>\n",
       "      <td>ACHX</td>\n",
       "      <td>CGAAGGCGCGGGCGGCGTCATTACCGATTGGTCGGGGCAGCCGTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Achromobacter xylosoxidans SOLR10</td>\n",
       "      <td>NZ_CP025774</td>\n",
       "      <td>complement(5049..5855)</td>\n",
       "      <td>338.906377</td>\n",
       "      <td>344.525203</td>\n",
       "      <td>239.579899</td>\n",
       "      <td>222.429542</td>\n",
       "      <td>237.845060</td>\n",
       "      <td>159.168702</td>\n",
       "      <td>260.879136</td>\n",
       "      <td>...</td>\n",
       "      <td>8.886514</td>\n",
       "      <td>10.508767</td>\n",
       "      <td>115.769842</td>\n",
       "      <td>95.915567</td>\n",
       "      <td>75.475714</td>\n",
       "      <td>208.814208</td>\n",
       "      <td>333.092076</td>\n",
       "      <td>332.654255</td>\n",
       "      <td>ACHX</td>\n",
       "      <td>AAGGAATACACGGCAACGGCGATGAGCTGCTTTTCTGCGGACGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96406</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>2000927..2003506</td>\n",
       "      <td>154.651786</td>\n",
       "      <td>166.363005</td>\n",
       "      <td>158.597735</td>\n",
       "      <td>103.096554</td>\n",
       "      <td>117.077900</td>\n",
       "      <td>113.354978</td>\n",
       "      <td>153.575746</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>186.116702</td>\n",
       "      <td>195.310415</td>\n",
       "      <td>191.215538</td>\n",
       "      <td>121.440741</td>\n",
       "      <td>149.782243</td>\n",
       "      <td>68.981909</td>\n",
       "      <td>SSUIS</td>\n",
       "      <td>TGGCCATTACGACTGTTGCGAACAAAGTTCGCCTTATCTCCAACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96407</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>complement(2003907..2004590)</td>\n",
       "      <td>51.843037</td>\n",
       "      <td>47.737511</td>\n",
       "      <td>50.150766</td>\n",
       "      <td>30.035985</td>\n",
       "      <td>27.269205</td>\n",
       "      <td>22.901773</td>\n",
       "      <td>142.767970</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69.873527</td>\n",
       "      <td>55.227042</td>\n",
       "      <td>67.758348</td>\n",
       "      <td>94.401511</td>\n",
       "      <td>108.957265</td>\n",
       "      <td>44.032987</td>\n",
       "      <td>SSUIS</td>\n",
       "      <td>ATATAACTTTTATCATCGGTGGTAGTTTGGGTCTCGATTCTTGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96408</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>complement(2004615..2005094)</td>\n",
       "      <td>72.526107</td>\n",
       "      <td>104.306461</td>\n",
       "      <td>90.287062</td>\n",
       "      <td>67.168967</td>\n",
       "      <td>49.922529</td>\n",
       "      <td>76.483851</td>\n",
       "      <td>292.305112</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>161.530080</td>\n",
       "      <td>151.730774</td>\n",
       "      <td>194.563257</td>\n",
       "      <td>245.576984</td>\n",
       "      <td>273.601934</td>\n",
       "      <td>308.030757</td>\n",
       "      <td>SSUIS</td>\n",
       "      <td>TTTTTTCATATAAGGAATCCTTTCCATTTCTTTAATTTTCCTTAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96409</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>2005268..2006464</td>\n",
       "      <td>133.774758</td>\n",
       "      <td>140.802925</td>\n",
       "      <td>145.617949</td>\n",
       "      <td>191.050633</td>\n",
       "      <td>234.493519</td>\n",
       "      <td>234.235125</td>\n",
       "      <td>136.907216</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180.352408</td>\n",
       "      <td>159.162080</td>\n",
       "      <td>162.029477</td>\n",
       "      <td>143.215049</td>\n",
       "      <td>132.640597</td>\n",
       "      <td>134.195769</td>\n",
       "      <td>SSUIS</td>\n",
       "      <td>TCCAACGGTAATCAATTTTATTTTCATAAAATAATTGTAACATATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96410</th>\n",
       "      <td>Streptococcus suis S10 -P1/7</td>\n",
       "      <td>NC_012925</td>\n",
       "      <td>2006525..2007289</td>\n",
       "      <td>322.297911</td>\n",
       "      <td>297.998136</td>\n",
       "      <td>289.298762</td>\n",
       "      <td>276.320634</td>\n",
       "      <td>306.466652</td>\n",
       "      <td>321.135201</td>\n",
       "      <td>448.980652</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>300.848584</td>\n",
       "      <td>373.308813</td>\n",
       "      <td>348.861950</td>\n",
       "      <td>313.318233</td>\n",
       "      <td>372.073690</td>\n",
       "      <td>286.332148</td>\n",
       "      <td>SSUIS</td>\n",
       "      <td>TTCAAAATAAGAGTGACCTCCAGAGCGCTCTCTACAAACATCAAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96411 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Species   Chromosome  \\\n",
       "0      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "1      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "2      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "3      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "4      Achromobacter xylosoxidans SOLR10  NZ_CP025774   \n",
       "...                                  ...          ...   \n",
       "96406       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "96407       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "96408       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "96409       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "96410       Streptococcus suis S10 -P1/7    NC_012925   \n",
       "\n",
       "                             Region  As_1 (GE) - TPM  As_2 (GE) - TPM  \\\n",
       "0                complement(1..400)      4579.764521      2874.021275   \n",
       "1             complement(980..2143)        14.339311        17.782257   \n",
       "2            complement(2164..3555)         8.624839        10.448941   \n",
       "3            complement(3599..4984)        21.127260        38.747719   \n",
       "4            complement(5049..5855)       338.906377       344.525203   \n",
       "...                             ...              ...              ...   \n",
       "96406              2000927..2003506       154.651786       166.363005   \n",
       "96407  complement(2003907..2004590)        51.843037        47.737511   \n",
       "96408  complement(2004615..2005094)        72.526107       104.306461   \n",
       "96409              2005268..2006464       133.774758       140.802925   \n",
       "96410              2006525..2007289       322.297911       297.998136   \n",
       "\n",
       "       As_3 (GE) - TPM  Bs_1 (GE) - TPM  Bs_2 (GE) - TPM  Bs_3 (GE) - TPM  \\\n",
       "0          2805.289041      3462.302060      3501.986142      3502.851956   \n",
       "1            23.076558        14.063192        17.260141        12.550914   \n",
       "2             3.816945         6.082623         6.523412         3.794404   \n",
       "3            23.426751        15.679651        17.443773        17.756848   \n",
       "4           239.579899       222.429542       237.845060       159.168702   \n",
       "...                ...              ...              ...              ...   \n",
       "96406       158.597735       103.096554       117.077900       113.354978   \n",
       "96407        50.150766        30.035985        27.269205        22.901773   \n",
       "96408        90.287062        67.168967        49.922529        76.483851   \n",
       "96409       145.617949       191.050633       234.493519       234.235125   \n",
       "96410       289.298762       276.320634       306.466652       321.135201   \n",
       "\n",
       "       Ctrl_1 (GE) - TPM  ...  Sp_2 (GE) - TPM  Sp_3 (GE) - TPM  \\\n",
       "0            3906.687837  ...      8965.167866      9056.547435   \n",
       "1              22.462166  ...        12.322023         9.714290   \n",
       "2               9.456728  ...         1.545564         3.046184   \n",
       "3              43.099754  ...         3.621928         3.824213   \n",
       "4             260.879136  ...         8.886514        10.508767   \n",
       "...                  ...  ...              ...              ...   \n",
       "96406         153.575746  ...              NaN              NaN   \n",
       "96407         142.767970  ...              NaN              NaN   \n",
       "96408         292.305112  ...              NaN              NaN   \n",
       "96409         136.907216  ...              NaN              NaN   \n",
       "96410         448.980652  ...              NaN              NaN   \n",
       "\n",
       "       Tm_1 (GE) - TPM  Tm_2 (GE) - TPM  Tm_3 (GE) - TPM  Vic_1 (GE) - TPM  \\\n",
       "0          3205.688636      3228.682954      3103.641082       3401.841238   \n",
       "1            30.958629        32.159798        33.681898         22.375684   \n",
       "2            10.307187        11.364142        15.717096          6.735852   \n",
       "3            46.222022        37.725002        48.744498         29.690886   \n",
       "4           115.769842        95.915567        75.475714        208.814208   \n",
       "...                ...              ...              ...               ...   \n",
       "96406       186.116702       195.310415       191.215538        121.440741   \n",
       "96407        69.873527        55.227042        67.758348         94.401511   \n",
       "96408       161.530080       151.730774       194.563257        245.576984   \n",
       "96409       180.352408       159.162080       162.029477        143.215049   \n",
       "96410       300.848584       373.308813       348.861950        313.318233   \n",
       "\n",
       "       Vic_2 (GE) - TPM  Vic_3 (GE) - TPM  Species ID  \\\n",
       "0           3832.566805       3408.199122        ACHX   \n",
       "1             22.529990         19.090084        ACHX   \n",
       "2              5.521991          4.961554        ACHX   \n",
       "3             27.566362         33.364651        ACHX   \n",
       "4            333.092076        332.654255        ACHX   \n",
       "...                 ...               ...         ...   \n",
       "96406        149.782243         68.981909       SSUIS   \n",
       "96407        108.957265         44.032987       SSUIS   \n",
       "96408        273.601934        308.030757       SSUIS   \n",
       "96409        132.640597        134.195769       SSUIS   \n",
       "96410        372.073690        286.332148       SSUIS   \n",
       "\n",
       "                                             upstream200  \n",
       "0      CCTTCCAAGCTTACGACGAGGGTTCGATTCCCTTCACCCGCTCCAA...  \n",
       "1      GCATTGCATCTCGGCCGCCAGTTGCGCACGGGCACCGTCTTCATGA...  \n",
       "2      GGAAGCGGCGCAGCGCCTGCGCCAGTCCGAAGCCGCGCGCGAGCTG...  \n",
       "3      CGAAGGCGCGGGCGGCGTCATTACCGATTGGTCGGGGCAGCCGTTA...  \n",
       "4      AAGGAATACACGGCAACGGCGATGAGCTGCTTTTCTGCGGACGGCG...  \n",
       "...                                                  ...  \n",
       "96406  TGGCCATTACGACTGTTGCGAACAAAGTTCGCCTTATCTCCAACCT...  \n",
       "96407  ATATAACTTTTATCATCGGTGGTAGTTTGGGTCTCGATTCTTGTAT...  \n",
       "96408  TTTTTTCATATAAGGAATCCTTTCCATTTCTTTAATTTTCCTTAAG...  \n",
       "96409  TCCAACGGTAATCAATTTTATTTTCATAAAATAATTGTAACATATC...  \n",
       "96410  TTCAAAATAAGAGTGACCTCCAGAGCGCTCTCTACAAACATCAAAT...  \n",
       "\n",
       "[96411 rows x 41 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8baa502a-5e70-4a5b-886d-07e81d586306",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df.to_csv('combined_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90090015-cbc6-49a9-8eec-1d6febf5a8f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
